{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generalizing mutagenesis pipeline\n",
    "\n",
    "Allowing for facile input of a transcript id/other information to automate process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samgould/opt/anaconda3/lib/python3.9/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import gffutils\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "import pandas as pd \n",
    "from pegg import prime\n",
    "import Bio.Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing human reference genome\n",
    "filepath = '/Users/samgould/Desktop/FSR Lab/reference files/GRCm39 (mouse)/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict, i = prime.genome_loader(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GRCh38 database\n",
    "#loading in genome annotations\n",
    "db = gffutils.FeatureDB('/Users/samgould/Desktop/FSR Lab/reference files/gencode_vm33_GRCm39.db', keep_order=True)\n",
    "\n",
    "#accessed genome annotations using gffutils package\n",
    "#canonical MANE transcript for CDK9 (CDK9-201)\n",
    "tx = 'ENST00000373264.5'\n",
    "\n",
    "\n",
    "def tx_processor(tx):\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    strand = db[tx].strand\n",
    "    chrom = db[tx].chrom\n",
    "    #print(chrom[3:])\n",
    "    #print(strand)\n",
    "    #print(db[tx].attributes)\n",
    "\n",
    "    #including 20 nt buffer on either side of exon for generating PAM sequences\n",
    "    start_end_cds_20 = []\n",
    "    buffer = 20\n",
    "    for k in start_end_cds:\n",
    "        h = []\n",
    "        for idx, j in enumerate(k):\n",
    "            if idx==0:\n",
    "                h.append(j-buffer)\n",
    "            if idx==1:\n",
    "                h.append(j+buffer)\n",
    "        start_end_cds_20.append(h)\n",
    "\n",
    "    return start_end_cds_20, start_end_cds, chrom, strand\n",
    "\n",
    "#tx_processor(tx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutagenize(tx, chrom_dict):\n",
    "    \n",
    "    start_end_cds_20, start_end_cds, chrom, strand = tx_processor(tx)\n",
    "    #chr9 = chrom_dict[9]\n",
    "    if chrom != 'chrX':\n",
    "        chr9 = chrom_dict[int(chrom[3:])]\n",
    "    else:\n",
    "        chr9 = chrom_dict[chrom[3:]]\n",
    "    #chr9 = chrom_dict[int(chrom[3:])]\n",
    "\n",
    "    plus_count = 0\n",
    "    minus_count = 0\n",
    "\n",
    "    plus_pam_loc_start = []\n",
    "    plus_pam_loc_end = []\n",
    "    PAM_plus = []\n",
    "    plus_protospacer = []\n",
    "    plus_proto_start = []\n",
    "    plus_proto_end = []\n",
    "    plus_ideal_start = []\n",
    "    plus_ideal_end = []\n",
    "    plus_ideal_window = []\n",
    "    plus_exon_list = []\n",
    "\n",
    "    minus_pam_loc_start = []\n",
    "    minus_pam_loc_end = []\n",
    "    PAM_minus = []\n",
    "    minus_protospacer = []\n",
    "    minus_proto_start = []\n",
    "    minus_proto_end = []\n",
    "    minus_ideal_start = []\n",
    "    minus_ideal_end = []\n",
    "    minus_ideal_window = []\n",
    "    minus_exon_list = []\n",
    "\n",
    "    #need to modify this to run it more generally!!!\n",
    "    #COME BACK TO THIS\n",
    "    #exon_list = [1,2,3,4,5,6,7]\n",
    "    if strand=='+':\n",
    "        exon_list = list(range(1, len(start_end_cds)+1))\n",
    "    elif strand=='-':\n",
    "        exon_list = list(range(1, len(start_end_cds)+1))[::-1]\n",
    "\n",
    "    for idx1, i in enumerate(start_end_cds_20):\n",
    "        s = i[0]-1\n",
    "        e = i[1]\n",
    "\n",
    "        sub_chrom=str(chr9[s:e]).upper()\n",
    "        #print(sub_chrom)\n",
    "        loc_list = list(range(s, e))\n",
    "\n",
    "        for idx, k in enumerate(sub_chrom):\n",
    "            #first deal with the plus strand\n",
    "            if k=='G':\n",
    "                \n",
    "                if idx-20 >=0:\n",
    "                    plus_count+=1\n",
    "                    plus_pam_loc_start.append(loc_list[idx])\n",
    "                    plus_pam_loc_end.append(loc_list[idx]+1)\n",
    "\n",
    "                    PAM_plus.append(str(chr9[loc_list[idx]-1:loc_list[idx]+1]).upper())\n",
    "                    \n",
    "                    proto_start = loc_list[idx]-21\n",
    "                    proto_end = loc_list[idx]-1\n",
    "                    proto = str(chr9[proto_start: proto_end]).upper()\n",
    "                    plus_protospacer.append(proto)\n",
    "                    plus_proto_start.append(proto_start+1)\n",
    "                    plus_proto_end.append(proto_end)\n",
    "\n",
    "                    #and ideal editing window\n",
    "                    plus_ideal_start.append(proto_start+3+1)\n",
    "                    plus_ideal_end.append(proto_start+8)\n",
    "                    plus_ideal_window.append(str(chr9[proto_start+3: proto_start+8]).upper())\n",
    "                    plus_exon_list.append(exon_list[idx1])\n",
    "\n",
    "            #then the minus strand\n",
    "            if k=='C':\n",
    "\n",
    "                if idx+20<=len(sub_chrom):\n",
    "                    minus_count+=1\n",
    "\n",
    "                    minus_pam_loc_end.append(loc_list[idx])\n",
    "                    minus_pam_loc_start.append(loc_list[idx]+1)\n",
    "\n",
    "                    PAM_minus.append(str(chr9[loc_list[idx]:loc_list[idx]+2].reverse_complement()).upper())\n",
    "\n",
    "                    proto_end = loc_list[idx]+2\n",
    "                    proto_start = loc_list[idx] + 22\n",
    "                    proto = str(chr9[proto_end: proto_start].reverse_complement()).upper()\n",
    "\n",
    "                    minus_protospacer.append(proto)\n",
    "                    minus_proto_start.append(proto_start)\n",
    "                    minus_proto_end.append(proto_end+1)\n",
    "\n",
    "                    #and ideal editing window\n",
    "                    minus_ideal_start.append(proto_start-3)\n",
    "                    minus_ideal_end.append(proto_start-8+1)\n",
    "                    minus_ideal_window.append(str(chr9[proto_start-8: proto_start-3].reverse_complement()).upper())\n",
    "                    minus_exon_list.append(exon_list[idx1])\n",
    "\n",
    "    col_names = ['protospacer', 'exon', 'proto_start', 'proto_end', 'PAM', 'PAM_start', 'PAM_end', 'ideal_start', 'ideal_end', 'ideal_window']\n",
    "\n",
    "    cols = [plus_protospacer, plus_exon_list, plus_proto_start,plus_proto_end,PAM_plus,plus_pam_loc_start,plus_pam_loc_end, plus_ideal_start, plus_ideal_end, plus_ideal_window ]\n",
    "    plus_df = pd.DataFrame(dict(zip(col_names, cols)))\n",
    "    plus_df['strand'] = '+'\n",
    "\n",
    "\n",
    "    cols_minus = [minus_protospacer, minus_exon_list, minus_proto_start,minus_proto_end,PAM_minus,minus_pam_loc_start,minus_pam_loc_end, minus_ideal_start, minus_ideal_end, minus_ideal_window ]\n",
    "\n",
    "    minus_df = pd.DataFrame(dict(zip(col_names, cols_minus)))\n",
    "    minus_df['strand'] = '-'\n",
    "    \n",
    "\n",
    "    #filter out guides whose ideal editing window doesn't hit in an exon\n",
    "    cds_hit = []\n",
    "    for i, val in plus_df.iterrows():\n",
    "        ideal_start = val['ideal_start']\n",
    "        ideal_end = val['ideal_end']-1\n",
    "        \n",
    "        hit=False\n",
    "        for k in start_end_cds:\n",
    "            exon_coords = list(range(k[0], k[1]+1))\n",
    "            if ideal_start in exon_coords:\n",
    "                hit=True\n",
    "            if ideal_end in exon_coords:\n",
    "                hit=True\n",
    "            \n",
    "        cds_hit.append(hit)\n",
    "\n",
    "    plus_df['CDS_hit'] = cds_hit\n",
    "\n",
    "    #and do the same for the minus strand\n",
    "    #filter out guides whose ideal editing window doesn't hit in an exon\n",
    "    cds_hit = []\n",
    "    for i, val in minus_df.iterrows():\n",
    "        ideal_start = val['ideal_start']-1\n",
    "        ideal_end = val['ideal_end']\n",
    "        \n",
    "        hit=False\n",
    "        for k in start_end_cds:\n",
    "            exon_coords = list(range(k[0], k[1]+1))\n",
    "            if ideal_start in exon_coords:\n",
    "                hit=True\n",
    "            if ideal_end in exon_coords:\n",
    "                hit=True\n",
    "            \n",
    "        cds_hit.append(hit)\n",
    "\n",
    "    minus_df['CDS_hit'] = cds_hit\n",
    "    \n",
    "    return pd.concat((plus_df, minus_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulated Mutagenesis \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'K2G_A5T_Y7R'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt_seq = 'KKGTACYW'\n",
    "mut_seq = 'KGGTTCRW'\n",
    "\n",
    "def hgvsp_simple(wt_seq, mut_seq):\n",
    "    if wt_seq==mut_seq:\n",
    "        return 'WT'\n",
    "    else:\n",
    "        pos_mutated = []\n",
    "\n",
    "        for i in range(len(wt_seq)):\n",
    "            if wt_seq[i] != mut_seq[i]:\n",
    "                pos_mutated.append(i + 1)\n",
    "        \n",
    "        hg = ''\n",
    "        for idx, pos in enumerate(pos_mutated):\n",
    "            if idx==0:\n",
    "                hg+= f'{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}'\n",
    "            else:\n",
    "                hg+= f'_{wt_seq[pos-1]}{pos}{mut_seq[pos-1]}'\n",
    "        \n",
    "\n",
    "        return hg\n",
    "\n",
    "hgvsp_simple(wt_seq, mut_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "def ABE_mutater(ideal_window):\n",
    "    #count the As\n",
    "    count_A = ideal_window.count('A')\n",
    "\n",
    "    #generate all possible combinatoric mutation possibilities\n",
    "    list_combs = []\n",
    "    for i in range(1, count_A+1):\n",
    "\n",
    "        list_combs+=list(itertools.combinations(range(count_A), i))\n",
    "\n",
    "    #index the As\n",
    "    l_key = []\n",
    "    counter=0\n",
    "    for i in ideal_window:\n",
    "        if i=='A':\n",
    "            l_key.append(counter)\n",
    "            counter+=1\n",
    "        else:\n",
    "            l_key.append('None')\n",
    "\n",
    "    #generate the mutated sequences\n",
    "    mutated_ideal = []\n",
    "    for i in list_combs:\n",
    "        len_combination = len(i)\n",
    "\n",
    "        mut_seq = ''\n",
    "        #for k in i:\n",
    "        for idx, base in enumerate(ideal_window):\n",
    "            if l_key[idx] in i:\n",
    "                mut_seq+='G'\n",
    "            else:\n",
    "                mut_seq+=base\n",
    "\n",
    "        mutated_ideal.append(mut_seq)\n",
    "\n",
    "    return mutated_ideal\n",
    "\n",
    "def CBE_mutater(ideal_window):\n",
    "    #count the As\n",
    "    count_C = ideal_window.count('C')\n",
    "\n",
    "    #generate all possible combinatoric mutation possibilities\n",
    "    list_combs = []\n",
    "    for i in range(1, count_C+1):\n",
    "\n",
    "        list_combs+=list(itertools.combinations(range(count_C), i))\n",
    "\n",
    "    #index the As\n",
    "    l_key = []\n",
    "    counter=0\n",
    "    for i in ideal_window:\n",
    "        if i=='C':\n",
    "            l_key.append(counter)\n",
    "            counter+=1\n",
    "        else:\n",
    "            l_key.append('None')\n",
    "\n",
    "    #generate the mutated sequences\n",
    "    mutated_ideal = []\n",
    "    for i in list_combs:\n",
    "        len_combination = len(i)\n",
    "\n",
    "        mut_seq = ''\n",
    "        #for k in i:\n",
    "        for idx, base in enumerate(ideal_window):\n",
    "            if l_key[idx] in i:\n",
    "                mut_seq+='T'\n",
    "            else:\n",
    "                mut_seq+=base\n",
    "\n",
    "        mutated_ideal.append(mut_seq)\n",
    "\n",
    "    return mutated_ideal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plus_mutate(plus_df):\n",
    "\n",
    "    cds_plus = plus_df[plus_df['CDS_hit']==True]\n",
    "    cds_plus = cds_plus.reset_index().drop(columns='index')\n",
    "\n",
    "\n",
    "    cds_plus['gRNA_id'] = [f'gRNA_{i+1}' for i in range(len(cds_plus))]\n",
    "\n",
    "    #perform combinatoric replacement\n",
    "    #ABE: A>G\n",
    "    #CBE: C>T\n",
    "    ABE_amenable = []\n",
    "    CBE_amenable = []\n",
    "\n",
    "    ABE_edits = []\n",
    "    ABE_edit_guides = []\n",
    "    CBE_edits = []\n",
    "    CBE_edit_guides = []\n",
    "\n",
    "    for i, val in cds_plus.iterrows():\n",
    "        ideal_window = val['ideal_window']\n",
    "        gRNA = val['gRNA_id']\n",
    "\n",
    "        #start with ABE:\n",
    "        if 'A' in ideal_window:\n",
    "            ABE_amenable.append(True)\n",
    "            ABE_mutations = ABE_mutater(ideal_window)\n",
    "            for k in ABE_mutations:\n",
    "                ABE_edits.append(k)\n",
    "                ABE_edit_guides.append(gRNA)\n",
    "\n",
    "        else:\n",
    "            ABE_amenable.append(False)\n",
    "\n",
    "        if 'C' in ideal_window:\n",
    "            CBE_amenable.append(True)\n",
    "            CBE_mutations = CBE_mutater(ideal_window)\n",
    "            for k in CBE_mutations:\n",
    "                CBE_edits.append(k)\n",
    "                CBE_edit_guides.append(gRNA)\n",
    "\n",
    "        else:\n",
    "            CBE_amenable.append(False)\n",
    "\n",
    "\n",
    "    cds_plus['ABE_amenable'] = ABE_amenable\n",
    "    cds_plus['CBE_amenable'] = CBE_amenable\n",
    "\n",
    "    CBE_edits_df = pd.DataFrame(dict(zip(['gRNA_id', 'edit'], [CBE_edit_guides, CBE_edits])))\n",
    "    ABE_edits_df = pd.DataFrame(dict(zip(['gRNA_id', 'edit'], [ABE_edit_guides, ABE_edits])))\n",
    "    CBE_edits_df['Editor'] = 'CBE'\n",
    "    ABE_edits_df['Editor'] = 'ABE'\n",
    "\n",
    "    return cds_plus, CBE_edits_df, ABE_edits_df\n",
    "\n",
    "def minus_mutate(minus_df, plus_df):\n",
    "    cds_minus = minus_df[minus_df['CDS_hit']==True]\n",
    "    cds_minus = cds_minus.reset_index().drop(columns='index')\n",
    "\n",
    "\n",
    "    cds_minus['gRNA_id'] = [f'gRNA_{i+1}' for i in range(len(plus_df), len(plus_df)+len(cds_minus))]\n",
    "\n",
    "    #perform combinatoric replacement\n",
    "    #ABE: A>G\n",
    "    #CBE: C>T\n",
    "    ABE_amenable = []\n",
    "    CBE_amenable = []\n",
    "\n",
    "    ABE_edits = []\n",
    "    ABE_edit_guides = []\n",
    "    CBE_edits = []\n",
    "    CBE_edit_guides = []\n",
    "\n",
    "    for i, val in cds_minus.iterrows():\n",
    "        ideal_window = val['ideal_window']\n",
    "        gRNA = val['gRNA_id']\n",
    "\n",
    "        #start with ABE:\n",
    "        if 'A' in ideal_window:\n",
    "            ABE_amenable.append(True)\n",
    "            ABE_mutations = ABE_mutater(ideal_window)\n",
    "            for k in ABE_mutations:\n",
    "                ABE_edits.append(k)\n",
    "                ABE_edit_guides.append(gRNA)\n",
    "\n",
    "        else:\n",
    "            ABE_amenable.append(False)\n",
    "\n",
    "        if 'C' in ideal_window:\n",
    "            CBE_amenable.append(True)\n",
    "            CBE_mutations = CBE_mutater(ideal_window)\n",
    "            for k in CBE_mutations:\n",
    "                CBE_edits.append(k)\n",
    "                CBE_edit_guides.append(gRNA)\n",
    "\n",
    "        else:\n",
    "            CBE_amenable.append(False)\n",
    "\n",
    "\n",
    "    cds_minus['ABE_amenable'] = ABE_amenable\n",
    "    cds_minus['CBE_amenable'] = CBE_amenable\n",
    "\n",
    "    CBE_edits_df = pd.DataFrame(dict(zip(['gRNA_id', 'edit'], [CBE_edit_guides, CBE_edits])))\n",
    "    ABE_edits_df = pd.DataFrame(dict(zip(['gRNA_id', 'edit'], [ABE_edit_guides, ABE_edits])))\n",
    "    CBE_edits_df['Editor'] = 'CBE'\n",
    "    ABE_edits_df['Editor'] = 'ABE'\n",
    "\n",
    "    return cds_minus, CBE_edits_df, ABE_edits_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_mutations_plus(input_df, WT_transcript_full, start_end_cds, strand, chrom_dict, chrom):\n",
    "\n",
    "    #chr9 = chrom_dict[int(chrom[3:])]\n",
    "    if chrom != 'chrX':\n",
    "        chr9 = chrom_dict[int(chrom[3:])]\n",
    "    else:\n",
    "        chr9 = chrom_dict[chrom[3:]]\n",
    "\n",
    "    hgvsp_holder = []\n",
    "    potential_splice_holder = []\n",
    "\n",
    "    wt_seq = str(Bio.Seq.Seq(WT_transcript_full).transcribe().translate())\n",
    "\n",
    "    for i, val in input_df.iterrows():\n",
    "        wt_ideal_seq = val['ideal_window']\n",
    "        mut_ideal_seq = val['edit']\n",
    "        ideal_start = val['ideal_start']\n",
    "        ideal_end = val['ideal_end']\n",
    "\n",
    "        mut_transcript = ''\n",
    "        wt_transcript = ''\n",
    "\n",
    "        potential_splice = False\n",
    "\n",
    "        for i in start_end_cds:\n",
    "            s = i[0]\n",
    "            e = i[1]\n",
    "   \n",
    "\n",
    "            contained=False\n",
    "            contained_fully = False\n",
    "            contained_start = False\n",
    "            contained_end = False\n",
    "\n",
    "            if ((ideal_end <= e) and (ideal_end>=s) and (ideal_start <= e) and (ideal_start>=s)):\n",
    "                contained=True\n",
    "                contained_fully=True\n",
    "\n",
    "            else:\n",
    "                if ((ideal_start <= e) and (ideal_start>=s)):\n",
    "                    contained=True\n",
    "                    contained_start = True\n",
    "                    potential_splice=True\n",
    "\n",
    "                if ((ideal_end <= e) and (ideal_end>=s)):\n",
    "                    contained=True\n",
    "                    contained_end=True\n",
    "                    potential_splice=True\n",
    "                    \n",
    "            if contained==False:\n",
    "                mut_transcript+=chr9[s-1:e]\n",
    "                wt_transcript+=chr9[s-1:e]\n",
    "            \n",
    "            elif contained==True:\n",
    "                #NEED TO DEAL WITH EDGE CASES WHERE IDEAL EDITING WINDOW OVERLAPS WITH BEGINNING/END OF EXON\n",
    "                #the first example (first set of parameters) has this precise issue...\n",
    "                if contained_fully == True:\n",
    "                    sub_tx_wt = chr9[s-1:ideal_start-1] + wt_ideal_seq + chr9[ideal_end:e]\n",
    "                    sub_tx_mut = chr9[s-1:ideal_start-1] + mut_ideal_seq + chr9[ideal_end:e]\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "\n",
    "                elif contained_start==True:\n",
    "                    #here the end needs to be cut off\n",
    "                    \n",
    "                    sub_tx_mut = chr9[s-1:ideal_start-1] + mut_ideal_seq[:e-ideal_end]\n",
    "                    sub_tx_wt = chr9[s-1:ideal_start-1] + wt_ideal_seq[:e-ideal_end]\n",
    "                    #print(val['gRNA_id'])\n",
    "                    #print(chr9[s-1:ideal_start])\n",
    "                    #print(wt_ideal_seq[:e-ideal_end-1])\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "\n",
    "\n",
    "                elif contained_end==True:\n",
    "                    #here the start needs to be cut off\n",
    "                    sub_tx_mut = mut_ideal_seq[s-ideal_start:] + chr9[ideal_end:e]\n",
    "                    sub_tx_wt = wt_ideal_seq[s-ideal_start:] + chr9[ideal_end:e]\n",
    "                    #print(sub_tx_wt)\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "\n",
    "\n",
    "        if strand=='+':\n",
    "            #print(sub_tx_wt)\n",
    "            #print(wt_transcript.upper() + '\\n'+ WT_transcript_full.upper())\n",
    "            \n",
    "            assert wt_transcript.upper()==WT_transcript_full.upper() , f'non-matching WT transcript | {val[\"gRNA_id\"]} {val[\"edit\"]} | {potential_splice} | {sub_tx_wt}'\n",
    "\n",
    "\n",
    "            mut_seq = str(Bio.Seq.Seq(mut_transcript).transcribe().translate())\n",
    "\n",
    "            hgvsp = hgvsp_simple(wt_seq, mut_seq)\n",
    "            hgvsp_holder.append(hgvsp)\n",
    "            \n",
    "            potential_splice_holder.append(potential_splice)\n",
    "            \n",
    "        elif strand=='-':\n",
    "\n",
    "            assert wt_transcript.reverse_complement().upper()==WT_transcript_full.upper(), f'non-matching WT transcript | {val[\"gRNA_id\"]} {val[\"edit\"]} | {potential_splice} | {sub_tx_wt}'\n",
    "\n",
    "\n",
    "            mut_seq = str(Bio.Seq.Seq(mut_transcript).reverse_complement().transcribe().translate())\n",
    "\n",
    "            hgvsp = hgvsp_simple(wt_seq, mut_seq)\n",
    "            hgvsp_holder.append(hgvsp)\n",
    "            \n",
    "            potential_splice_holder.append(potential_splice)\n",
    "\n",
    "    return hgvsp_holder, potential_splice_holder\n",
    "\n",
    "\n",
    "def simulate_mutations_minus(input_df, WT_transcript_full, start_end_cds, strand, chrom_dict, chrom):\n",
    "\n",
    "    #chr9 = chrom_dict[int(chrom[3:])]\n",
    "    if chrom != 'chrX':\n",
    "        chr9 = chrom_dict[int(chrom[3:])]\n",
    "    else:\n",
    "        chr9 = chrom_dict[chrom[3:]]\n",
    "    \n",
    "    hgvsp_holder = []\n",
    "    potential_splice_holder = []\n",
    "\n",
    "    wt_seq = str(Bio.Seq.Seq(WT_transcript_full).transcribe().translate())\n",
    "\n",
    "    for i, val in input_df.iterrows():\n",
    "        wt_ideal_seq = val['ideal_window']\n",
    "        mut_ideal_seq = val['edit']\n",
    "        ideal_start = val['ideal_end'] #swapped these\n",
    "        ideal_end = val['ideal_start']\n",
    "\n",
    "        wt_ideal_seq = str(Bio.Seq.Seq(wt_ideal_seq).reverse_complement())\n",
    "        mut_ideal_seq = str(Bio.Seq.Seq(mut_ideal_seq).reverse_complement())\n",
    "\n",
    "        mut_transcript = ''\n",
    "        wt_transcript = ''\n",
    "\n",
    "        potential_splice = False\n",
    "\n",
    "        for i in start_end_cds:\n",
    "            s = i[0]\n",
    "            e = i[1]\n",
    "\n",
    "            contained=False\n",
    "            contained_fully = False\n",
    "            contained_start = False\n",
    "            contained_end = False\n",
    "\n",
    "            if ((ideal_end <= e) and (ideal_end>=s) and (ideal_start <= e) and (ideal_start>=s)):\n",
    "                contained=True\n",
    "                contained_fully=True\n",
    "\n",
    "            else:\n",
    "                if ((ideal_start <= e) and (ideal_start>=s)):\n",
    "                    contained=True\n",
    "                    contained_start = True\n",
    "                    potential_splice=True\n",
    "\n",
    "                if ((ideal_end <= e) and (ideal_end>=s)):\n",
    "                    contained=True\n",
    "                    contained_end=True\n",
    "                    potential_splice=True\n",
    "                    \n",
    "            if contained==False:\n",
    "                mut_transcript+=chr9[s-1:e]\n",
    "                wt_transcript+=chr9[s-1:e]\n",
    "            \n",
    "            elif contained==True:\n",
    "                #NEED TO DEAL WITH EDGE CASES WHERE IDEAL EDITING WINDOW OVERLAPS WITH BEGINNING/END OF EXON\n",
    "                #the first example (first set of parameters) has this precise issue...\n",
    "                if contained_fully == True:\n",
    "                    sub_tx_wt = chr9[s-1:ideal_start-1] + wt_ideal_seq + chr9[ideal_end:e]\n",
    "                    sub_tx_mut = chr9[s-1:ideal_start-1] + mut_ideal_seq + chr9[ideal_end:e]\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "\n",
    "                elif contained_start==True:\n",
    "                    #here the end needs to be cut off\n",
    "                    \n",
    "                    sub_tx_mut = chr9[s-1:ideal_start-1] + mut_ideal_seq[:e-ideal_end]\n",
    "                    sub_tx_wt = chr9[s-1:ideal_start-1] + wt_ideal_seq[:e-ideal_end]\n",
    "                    #print(val['gRNA_id'])\n",
    "                    #print(chr9[s-1:ideal_start])\n",
    "                    #print(wt_ideal_seq[:e-ideal_end-1])\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "\n",
    "\n",
    "                elif contained_end==True:\n",
    "                    #here the start needs to be cut off\n",
    "                    sub_tx_mut = mut_ideal_seq[s-ideal_start:] + chr9[ideal_end:e]\n",
    "                    sub_tx_wt = wt_ideal_seq[s-ideal_start:] + chr9[ideal_end:e]\n",
    "                    #print(sub_tx_wt)\n",
    "                    mut_transcript+=sub_tx_mut\n",
    "                    wt_transcript +=sub_tx_wt\n",
    "        \n",
    "        if strand=='+':\n",
    "            assert wt_transcript.upper()==WT_transcript_full.upper(), f'non-matching WT transcript | {val[\"gRNA_id\"]} {val[\"edit\"]} | {potential_splice} | {sub_tx_wt}'\n",
    "\n",
    "\n",
    "            mut_seq = str(Bio.Seq.Seq(mut_transcript).transcribe().translate())\n",
    "\n",
    "            hgvsp = hgvsp_simple(wt_seq, mut_seq)\n",
    "            hgvsp_holder.append(hgvsp)\n",
    "            \n",
    "            potential_splice_holder.append(potential_splice)\n",
    "        \n",
    "        elif strand=='-':\n",
    "\n",
    "            assert wt_transcript.reverse_complement().upper()==WT_transcript_full.upper(), f'non-matching WT transcript | {val[\"gRNA_id\"]} {val[\"edit\"]} | {potential_splice} | {sub_tx_wt}'\n",
    "\n",
    "\n",
    "            mut_seq = str(Bio.Seq.Seq(mut_transcript).reverse_complement().transcribe().translate())\n",
    "\n",
    "            hgvsp = hgvsp_simple(wt_seq, mut_seq)\n",
    "            hgvsp_holder.append(hgvsp)\n",
    "            \n",
    "            potential_splice_holder.append(potential_splice)\n",
    "\n",
    "    return hgvsp_holder, potential_splice_holder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sensor_maker(guide_dataframe, chr9):\n",
    "    \"\"\" \n",
    "    function for making sensors in reverse complement orientation\n",
    "    42 nt sensor (10 nt on either side of protospacer + PAM sequence)\n",
    "    \"\"\"\n",
    "\n",
    "    sensors_rc = []\n",
    "\n",
    "    for i, val in guide_dataframe.iterrows():\n",
    "        strand = val['strand']\n",
    "        proto = val['protospacer']\n",
    "        pam = val['PAM']\n",
    "        proto_pam = proto+pam\n",
    "\n",
    "        if strand == '+':\n",
    "            s = val['proto_start']\n",
    "            e = val['proto_end']\n",
    "\n",
    "            sensor = str(chr9[s-11:e+10+2]).upper()\n",
    "\n",
    "        elif strand=='-':\n",
    "            s = val['proto_end']\n",
    "            e = val['proto_start']\n",
    "\n",
    "            sensor = str(Bio.Seq.Seq(chr9[s-3-10:e+10]).reverse_complement()).upper()\n",
    "\n",
    "        assert proto_pam == sensor[10:-10], print(f'{strand} | {sensor} | {proto_pam} | {sensor[10:-10]}')\n",
    "\n",
    "        \n",
    "        sensor_rc = str(Bio.Seq.Seq(sensor).reverse_complement())\n",
    "\n",
    "        assert len(sensor_rc)==42\n",
    "\n",
    "        sensors_rc.append(sensor_rc)\n",
    "\n",
    "    return sensors_rc\n",
    "\n",
    "def sensor_alt_maker(guides, edit_outcomes):\n",
    "\n",
    "    sensor_wt_list = []\n",
    "    sensor_alt_list = []\n",
    "\n",
    "    for i, val in edit_outcomes.iterrows():\n",
    "        grna = val['gRNA_id']\n",
    "        ideal_window = val['ideal_window']\n",
    "        edit = val['edit']\n",
    "\n",
    "        sensor = guides.loc[guides['gRNA_id']==grna, 'sensor_wt'].values[0]\n",
    "\n",
    "        sensor_wt_list.append(sensor)\n",
    "\n",
    "        sensor_rc = str(Bio.Seq.Seq(sensor).reverse_complement())\n",
    "        assert ideal_window == sensor_rc[13:18]\n",
    "\n",
    "        sensor_alt = sensor_rc[:13] + edit + sensor_rc[18:]\n",
    "        sensor_alt_rc = str(Bio.Seq.Seq(sensor_alt).reverse_complement())\n",
    "        sensor_alt_list.append(sensor_alt_rc)\n",
    "\n",
    "        assert len(sensor_alt_rc)==42\n",
    "        assert sensor_alt_rc != sensor\n",
    "        \n",
    "    edit_outcomes['sensor_wt'] = sensor_wt_list\n",
    "    edit_outcomes['sensor_alt'] = sensor_alt_list\n",
    "\n",
    "    return edit_outcomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutation_simulator(gene, cdks, chrom_dict):\n",
    "\n",
    "    tx = cdks.loc[cdks['Gene']==gene, 'Transcript ID'].values[0]\n",
    "    protein = cdks.loc[cdks['Gene']==gene, 'Protein'].values[0]\n",
    "\n",
    "    #--------check that it's the correct transcript and get the WT transcript sequence-------\n",
    "    start_end_cds_20, start_end_cds, chrom, strand = tx_processor(tx)\n",
    "    \n",
    "    if chrom != 'chrX':\n",
    "        chr9 = chrom_dict[int(chrom[3:])]\n",
    "    else:\n",
    "        chr9 = chrom_dict[chrom[3:]]\n",
    "    \n",
    "    transcript = ''\n",
    "    for i in start_end_cds:\n",
    "        s = i[0]-1\n",
    "        e = i[1]\n",
    "        transcript+=chr9[s:e]\n",
    "    \n",
    "    if strand == '+':\n",
    "        WT_transcript_full = str(transcript).upper()\n",
    "    elif strand =='-':\n",
    "        WT_transcript_full = str(transcript.reverse_complement()).upper()\n",
    "\n",
    "    assert str(Bio.Seq.Seq(WT_transcript_full).transcribe().translate()) == protein\n",
    "\n",
    "    #-----------perform the mutagenesis-----------\n",
    "    out = mutagenize(tx, chrom_dict)\n",
    "\n",
    "    #-------select only hits on the CDS--------\n",
    "    out = out[out['CDS_hit']==True]\n",
    "    plus_df = out[out['strand']=='+']\n",
    "    minus_df = out[out['strand']=='-']\n",
    "\n",
    "    #--------and then simulate the mutations--------\n",
    "\n",
    "    #---------plus sequence-------------\n",
    "    cds_plus, CBE_edits_df, ABE_edits_df  = plus_mutate(plus_df)\n",
    "\n",
    "    start_end_cds_20, start_end_cds, chrom, strand = tx_processor(tx)\n",
    "    \n",
    "    ABE1 = pd.merge(ABE_edits_df, cds_plus, on='gRNA_id')\n",
    "    #return ABE1\n",
    "    hgvsp_holder, potential_splice_holder = simulate_mutations_plus(ABE1, WT_transcript_full, start_end_cds,strand, chrom_dict, chrom)\n",
    "    ABE1['HGVSp'] = hgvsp_holder\n",
    "    ABE1['potential_splice'] = potential_splice_holder\n",
    "\n",
    "    CBE1 = pd.merge(CBE_edits_df, cds_plus, on='gRNA_id')\n",
    "    hgvsp_holder_c, potential_splice_holder_c = simulate_mutations_plus(CBE1, WT_transcript_full, start_end_cds,strand, chrom_dict, chrom)\n",
    "    CBE1['HGVSp'] = hgvsp_holder_c\n",
    "    CBE1['potential_splice'] = potential_splice_holder_c\n",
    "\n",
    "    #---------minus sequence-------------\n",
    "    cds_minus, CBE_edits_df, ABE_edits_df = minus_mutate(minus_df, plus_df)\n",
    "    \n",
    "    ABE1_minus = pd.merge(ABE_edits_df, cds_minus, on='gRNA_id')\n",
    "    hgvsp_holder, potential_splice_holder = simulate_mutations_minus(ABE1_minus, WT_transcript_full, start_end_cds,strand, chrom_dict, chrom)\n",
    "    ABE1_minus['HGVSp'] = hgvsp_holder\n",
    "    ABE1_minus['potential_splice'] = potential_splice_holder\n",
    "\n",
    "    CBE1_minus = pd.merge(CBE_edits_df, cds_minus, on='gRNA_id')\n",
    "    hgvsp_holder_c, potential_splice_holder_c = simulate_mutations_minus(CBE1_minus, WT_transcript_full, start_end_cds, strand, chrom_dict, chrom)\n",
    "    CBE1_minus['HGVSp'] = hgvsp_holder_c\n",
    "    CBE1_minus['potential_splice'] = potential_splice_holder_c\n",
    "\n",
    "    guides = pd.concat((cds_plus, cds_minus))\n",
    "\n",
    "    if chrom != 'chrX':\n",
    "        guides['chrom'] = int(chrom[3:])\n",
    "    else:\n",
    "        guides['chrom'] = chrom[3:]\n",
    "\n",
    "    sensor_rc = sensor_maker(guides, chr9)\n",
    "    guides['sensor_wt'] = sensor_rc\n",
    "    guides['proto_G+19'] = [f'G{i[1:]}' for i in guides['protospacer']]\n",
    "\n",
    "    cols = ['gRNA_id','protospacer', 'proto_G+19', 'chrom', 'exon', 'proto_start', 'proto_end', 'PAM', 'PAM_start',\n",
    "    'PAM_end', 'ideal_start', 'ideal_end', 'ideal_window', 'strand','sensor_wt',\n",
    "    'CDS_hit', 'ABE_amenable', 'CBE_amenable', ]\n",
    "\n",
    "    guides = guides[cols]\n",
    "    \n",
    "    edit_outcomes = pd.concat((ABE1, CBE1, ABE1_minus, CBE1_minus)) \n",
    "\n",
    "    #------and get more info from the edit_outcomes---------\n",
    "    complex_list = []\n",
    "    mut_aa_list = []\n",
    "    wt_aa_list = []\n",
    "    codon_list = []\n",
    "    for i, val in edit_outcomes.iterrows():\n",
    "        hg = val['HGVSp']\n",
    "        if '_' in hg:\n",
    "            complex_list.append(True)\n",
    "            mut_aa_list.append(None)\n",
    "            wt_aa_list.append(None)\n",
    "            codon_list.append(None)\n",
    "        else:\n",
    "            complex_list.append(False)\n",
    "            if hg !='WT':\n",
    "                mut_aa_list.append(hg[-1])\n",
    "                wt_aa_list.append(hg[0])\n",
    "                codon_list.append(int(hg[1:-1]))\n",
    "            elif hg=='WT':\n",
    "                mut_aa_list.append(None)\n",
    "                wt_aa_list.append(None)\n",
    "                codon_list.append(None)\n",
    "\n",
    "\n",
    "    edit_outcomes['Complex'] = complex_list\n",
    "    edit_outcomes['MUT_AA'] = mut_aa_list \n",
    "    edit_outcomes['WT_AA'] = wt_aa_list\n",
    "    edit_outcomes['Codon'] = codon_list\n",
    "\n",
    "    #-------determine sensor wt and alt for each-----------\n",
    "    edit_outcomes = sensor_alt_maker(guides, edit_outcomes)\n",
    "    \n",
    "\n",
    "    return guides, edit_outcomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating gRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Men1</td>\n",
       "      <td>ENSMUST00000113504.10</td>\n",
       "      <td>MGLKAAQKTLFPLRSIDDVVRLFAAELGREEPDLVLLSLVLGFVEH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Kdm6a</td>\n",
       "      <td>ENSMUST00000044484.13</td>\n",
       "      <td>MKSCGVSLATAAAAAAAAAFGDEEKKMAAGKASGESEEASPSLTAE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kmt2a</td>\n",
       "      <td>ENSMUST00000002095.11</td>\n",
       "      <td>MAHSCRWRFPARPGTTGGGGGGGRRGLGGAPRQRVPALLLPPGPQA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Kmt2b</td>\n",
       "      <td>ENSMUST00000108154.9</td>\n",
       "      <td>MAAAAGGGSCPGPGSARGRFPGRPRGSGGGGGRGGRGNGAERVRVA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Kmt2c</td>\n",
       "      <td>ENSMUST00000045291.14</td>\n",
       "      <td>MSSEEDRSAEQQQPPPAPPEEPGAPAPSPAAADKRPRGRPRKDGAS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>ENSMUST00000023741.16</td>\n",
       "      <td>MDSQKPPAEDKDSDPAADGLAAPEKPGATEPDLPILCIGEVSVPGS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene          Transcript ID  \\\n",
       "0   Men1  ENSMUST00000113504.10   \n",
       "1  Kdm6a  ENSMUST00000044484.13   \n",
       "2  Kmt2a  ENSMUST00000002095.11   \n",
       "3  Kmt2b   ENSMUST00000108154.9   \n",
       "4  Kmt2c  ENSMUST00000045291.14   \n",
       "5  Kmt2d  ENSMUST00000023741.16   \n",
       "\n",
       "                                             Protein  \n",
       "0  MGLKAAQKTLFPLRSIDDVVRLFAAELGREEPDLVLLSLVLGFVEH...  \n",
       "1  MKSCGVSLATAAAAAAAAAFGDEEKKMAAGKASGESEEASPSLTAE...  \n",
       "2  MAHSCRWRFPARPGTTGGGGGGGRRGLGGAPRQRVPALLLPPGPQA...  \n",
       "3  MAAAAGGGSCPGPGSARGRFPGRPRGSGGGGGRGGRGNGAERVRVA...  \n",
       "4  MSSEEDRSAEQQQPPPAPPEEPGAPAPSPAAADKRPRGRPRKDGAS...  \n",
       "5  MDSQKPPAEDKDSDPAADGLAAPEKPGATEPDLPILCIGEVSVPGS...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fgfr2_3b = pd.read_excel('2025_01_14_NG_Tiling_Libraries_YMSF_FSR.xlsx')\n",
    "fgfr2_3b = fgfr2_3b[['Gene', 'Transcript ID', 'Protein']]\n",
    "\n",
    "#SELECT ONLY THE HUMAN TRANSCRIPTS FIRST\n",
    "fgfr2_3b = fgfr2_3b[6:].reset_index(drop=True)\n",
    "fgfr2_3b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Men1 : 1129 guides : 4634 edits\n",
      "Kdm6a : 1903 guides : 7833 edits\n",
      "Kmt2a : 6320 guides : 27371 edits\n",
      "Kmt2b : 5029 guides : 21295 edits\n",
      "Kmt2c : 6975 guides : 29075 edits\n",
      "Kmt2d : 10243 guides : 42657 edits\n"
     ]
    }
   ],
   "source": [
    "guide_dfs = []\n",
    "edit_dfs = []\n",
    "names = []\n",
    "\n",
    "for i in list(fgfr2_3b['Gene']):\n",
    "    names.append(i)\n",
    "    guides, edit_outcomes = mutation_simulator(i, fgfr2_3b, chrom_dict)\n",
    "    guide_dfs.append(guides)\n",
    "    edit_dfs.append(edit_outcomes)\n",
    "    print(f'{i} : {len(guides)} guides : {len(edit_outcomes)} edits')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "guide_dict = dict(zip(names, guide_dfs))\n",
    "edit_dict = dict(zip(names, edit_dfs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating Controls\n",
    "\n",
    "- Intron targeting controls\n",
    "- Non-targeting controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pegg import library\n",
    "from pegg import base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#and also generate 100 non-targeting guides for prime editing\n",
    "num_guides = 1000\n",
    "nontarget = library.nontargeting_guides(num_guides, edit_type='base')\n",
    "\n",
    "nontarget['proto_G+19'] = [f'G{i[2:]}' for i in nontarget['Protospacer']]\n",
    "\n",
    "rand_seqs = np.random.choice(['A','T','C','G'], size=(1000,42))\n",
    "rand_sensors_42 = [\"\".join(i) for i in rand_seqs]\n",
    "nontarget['sensor_seq_random'] = rand_sensors_42\n",
    "\n",
    "RE_sites_polyT = ['CGTCTC', 'GAATTC', 'GAGACG', 'TTTT']\n",
    "\n",
    "filt = []\n",
    "for i, val in nontarget.iterrows():\n",
    "    p= val['Protospacer']\n",
    "    s = val['sensor_seq_random']\n",
    "    \n",
    "    c = 0\n",
    "    for k in RE_sites_polyT:\n",
    "        if k in p:\n",
    "            c+=1\n",
    "        if k in s:\n",
    "            c+=1\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "    if c>0:\n",
    "        filt.append(True)\n",
    "    else:\n",
    "        filt.append(False)\n",
    "\n",
    "#and then generating \n",
    "\n",
    "nontarget['polyT_or_RE_site'] = filt\n",
    "nontarget = nontarget[nontarget['polyT_or_RE_site']==False].reset_index().drop(columns=['index'])\n",
    "#nontarget.to_csv('non_targeting_guides.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>classification</th>\n",
       "      <th>proto_G+19</th>\n",
       "      <th>sensor_seq_random</th>\n",
       "      <th>polyT_or_RE_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GACGGAGGCTAAGCGTCGCAA</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GCGGAGGCTAAGCGTCGCAA</td>\n",
       "      <td>AACCGCCGAAACATATGGTTCAAACGAGAAGGGAGACCAACC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCGCTTCCGCGGCCCGTTCAA</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GGCTTCCGCGGCCCGTTCAA</td>\n",
       "      <td>TTGGTTCACGGGTGCTTCGCAGACAAATGTTCCACGGCCTTC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATCGTTTCCGCTTAACGGCG</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GTCGTTTCCGCTTAACGGCG</td>\n",
       "      <td>CTCTTTATTGCAGAGACCCGCAAGCACCGTCAAGGCGGCACC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GGTAGGCGCGCCGCTCTCTAC</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GTAGGCGCGCCGCTCTCTAC</td>\n",
       "      <td>GGGTGGTAGGCTTTAGATCCAAAGAGCCGAGATGATCAGCAC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCCATATCGGGGCGAGACATG</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GCATATCGGGGCGAGACATG</td>\n",
       "      <td>AAATGATAGCTAGCGATTAGCACAAAGCCTGGGGGTGGCGGC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>820</th>\n",
       "      <td>GGACTGAAATCCAAGGACTGT</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GACTGAAATCCAAGGACTGT</td>\n",
       "      <td>ACTACTCAAATTTAGGGACACGATCTCATGTGGTCATTCTTA</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>GTAAACAAAAAGGAAATAGTT</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GAAACAAAAAGGAAATAGTT</td>\n",
       "      <td>CGCGGAACGGAACGTCTGGAGGCTGTAGGTTCTCTCCTCATG</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>GTTTCCCATGATCATTTAGTG</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GTTCCCATGATCATTTAGTG</td>\n",
       "      <td>TTTAAGGGATACCCGTGATGGGTGGCCATCATCCCGTGTGGT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>GCTGAGTGAAAAATAAAAGTT</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GTGAGTGAAAAATAAAAGTT</td>\n",
       "      <td>GAACTCGCCATGTCGAAGCTCCTCGTTCAGATGTCCTTCAAT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>GGACAATCATGGTGAAAGCGG</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>GACAATCATGGTGAAAGCGG</td>\n",
       "      <td>TTCGGATACGGTGGATTCTTCCGGCAGAACCCGTGGCGCTAT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>825 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Protospacer         classification            proto_G+19  \\\n",
       "0    GACGGAGGCTAAGCGTCGCAA  non-targeting control  GCGGAGGCTAAGCGTCGCAA   \n",
       "1    GCGCTTCCGCGGCCCGTTCAA  non-targeting control  GGCTTCCGCGGCCCGTTCAA   \n",
       "2    GATCGTTTCCGCTTAACGGCG  non-targeting control  GTCGTTTCCGCTTAACGGCG   \n",
       "3    GGTAGGCGCGCCGCTCTCTAC  non-targeting control  GTAGGCGCGCCGCTCTCTAC   \n",
       "4    GCCATATCGGGGCGAGACATG  non-targeting control  GCATATCGGGGCGAGACATG   \n",
       "..                     ...                    ...                   ...   \n",
       "820  GGACTGAAATCCAAGGACTGT  non-targeting control  GACTGAAATCCAAGGACTGT   \n",
       "821  GTAAACAAAAAGGAAATAGTT  non-targeting control  GAAACAAAAAGGAAATAGTT   \n",
       "822  GTTTCCCATGATCATTTAGTG  non-targeting control  GTTCCCATGATCATTTAGTG   \n",
       "823  GCTGAGTGAAAAATAAAAGTT  non-targeting control  GTGAGTGAAAAATAAAAGTT   \n",
       "824  GGACAATCATGGTGAAAGCGG  non-targeting control  GACAATCATGGTGAAAGCGG   \n",
       "\n",
       "                              sensor_seq_random  polyT_or_RE_site  \n",
       "0    AACCGCCGAAACATATGGTTCAAACGAGAAGGGAGACCAACC             False  \n",
       "1    TTGGTTCACGGGTGCTTCGCAGACAAATGTTCCACGGCCTTC             False  \n",
       "2    CTCTTTATTGCAGAGACCCGCAAGCACCGTCAAGGCGGCACC             False  \n",
       "3    GGGTGGTAGGCTTTAGATCCAAAGAGCCGAGATGATCAGCAC             False  \n",
       "4    AAATGATAGCTAGCGATTAGCACAAAGCCTGGGGGTGGCGGC             False  \n",
       "..                                          ...               ...  \n",
       "820  ACTACTCAAATTTAGGGACACGATCTCATGTGGTCATTCTTA             False  \n",
       "821  CGCGGAACGGAACGTCTGGAGGCTGTAGGTTCTCTCCTCATG             False  \n",
       "822  TTTAAGGGATACCCGTGATGGGTGGCCATCATCCCGTGTGGT             False  \n",
       "823  GAACTCGCCATGTCGAAGCTCCTCGTTCAGATGTCCTTCAAT             False  \n",
       "824  TTCGGATACGGTGGATTCTTCCGGCAGAACCCGTGGCGCTAT             False  \n",
       "\n",
       "[825 rows x 5 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nontarget"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intron targeting controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "tx = 'ENST00000373264.5'\n",
    "\n",
    "\n",
    "def tx_processor(tx):\n",
    "    cds = list(db.children(tx, order_by='+end', featuretype=['CDS']))\n",
    "    start_end_cds = [[i.start, i.end] for i in cds]\n",
    "    strand = db[tx].strand\n",
    "    chrom = db[tx].chrom\n",
    "    #print(chrom[3:])\n",
    "    #print(strand)\n",
    "    #print(db[tx].attributes)\n",
    "\n",
    "    #including 20 nt buffer on either side of exon for generating PAM sequences\n",
    "    start_end_cds_20 = []\n",
    "    buffer = 20\n",
    "    for k in start_end_cds:\n",
    "        h = []\n",
    "        for idx, j in enumerate(k):\n",
    "            if idx==0:\n",
    "                h.append(j-buffer)\n",
    "            if idx==1:\n",
    "                h.append(j+buffer)\n",
    "        start_end_cds_20.append(h)\n",
    "\n",
    "    return start_end_cds_20, start_end_cds, chrom, strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutagenize_intron(tx, chrom_dict):\n",
    "    \n",
    "    start_end_cds_20, start_end_cds, chrom, strand = tx_processor(tx)\n",
    "\n",
    "    intron_20 = []\n",
    "    intron_lens = []\n",
    "    for i, val in enumerate(start_end_cds_20):\n",
    "            \n",
    "        if i<len(start_end_cds_20)-1:\n",
    "            intron_20.append([start_end_cds_20[i][1], start_end_cds_20[i+1][0]])\n",
    "\n",
    "            intron_lens.append(start_end_cds_20[i+1][0] - start_end_cds_20[i][1])\n",
    "            \n",
    "        if i==len(start_end_cds_20)-1:\n",
    "            break\n",
    "\n",
    "    #print(intron_lens)\n",
    "\n",
    "    intron_20_new = []\n",
    "    intron_lens_new = []\n",
    "    #deal with the case of very large introns that will take forever to process\n",
    "\n",
    "    for i, val in enumerate(intron_lens):\n",
    "        if val>1000:\n",
    "            rand = np.random.randint(intron_20[i][0], intron_20[i][1]-1000)\n",
    "            intron_20_new.append([rand, rand+1000])\n",
    "            intron_lens_new.append(1000)\n",
    "        else:\n",
    "            intron_lens_new.append(val)\n",
    "            intron_20_new.append(intron_20[i])\n",
    "\n",
    "    start_end_cds_20 = intron_20_new\n",
    "\n",
    "\n",
    "    #chr9 = chrom_dict[9]\n",
    "\n",
    "    if chrom != 'chrX':\n",
    "        chr9 = chrom_dict[int(chrom[3:])]\n",
    "    else:\n",
    "        chr9 = chrom_dict[chrom[3:]]\n",
    "    #chr9 = chrom_dict[int(chrom[3:])]\n",
    "\n",
    "    plus_count = 0\n",
    "    minus_count = 0\n",
    "\n",
    "    plus_pam_loc_start = []\n",
    "    plus_pam_loc_end = []\n",
    "    PAM_plus = []\n",
    "    plus_protospacer = []\n",
    "    plus_proto_start = []\n",
    "    plus_proto_end = []\n",
    "    plus_ideal_start = []\n",
    "    plus_ideal_end = []\n",
    "    plus_ideal_window = []\n",
    "    plus_exon_list = []\n",
    "\n",
    "    minus_pam_loc_start = []\n",
    "    minus_pam_loc_end = []\n",
    "    PAM_minus = []\n",
    "    minus_protospacer = []\n",
    "    minus_proto_start = []\n",
    "    minus_proto_end = []\n",
    "    minus_ideal_start = []\n",
    "    minus_ideal_end = []\n",
    "    minus_ideal_window = []\n",
    "    minus_exon_list = []\n",
    "\n",
    "    #need to modify this to run it more generally!!!\n",
    "    #COME BACK TO THIS\n",
    "    #exon_list = [1,2,3,4,5,6,7]\n",
    "    if strand=='+':\n",
    "        exon_list = [f'intron{i}_{i+1}' for i in range(1, len(start_end_cds_20)+1)]\n",
    "    elif strand=='-':\n",
    "        exon_list = [f'intron{i}_{i+1}' for i in range(1, len(start_end_cds_20)+1)]\n",
    "\n",
    "    for idx1, i in enumerate(start_end_cds_20):\n",
    "        s = i[0]-1\n",
    "        e = i[1]\n",
    "\n",
    "        sub_chrom=str(chr9[s:e]).upper()\n",
    "        #print(sub_chrom)\n",
    "        loc_list = list(range(s, e))\n",
    "\n",
    "        for idx, k in enumerate(sub_chrom):\n",
    "            #first deal with the plus strand\n",
    "            if k=='G':\n",
    "                \n",
    "                if idx-20 >=0:\n",
    "                    plus_count+=1\n",
    "                    plus_pam_loc_start.append(loc_list[idx])\n",
    "                    plus_pam_loc_end.append(loc_list[idx]+1)\n",
    "\n",
    "                    PAM_plus.append(str(chr9[loc_list[idx]-1:loc_list[idx]+1]).upper())\n",
    "                    \n",
    "                    proto_start = loc_list[idx]-21\n",
    "                    proto_end = loc_list[idx]-1\n",
    "                    proto = str(chr9[proto_start: proto_end]).upper()\n",
    "                    plus_protospacer.append(proto)\n",
    "                    plus_proto_start.append(proto_start+1)\n",
    "                    plus_proto_end.append(proto_end)\n",
    "\n",
    "                    #and ideal editing window\n",
    "                    plus_ideal_start.append(proto_start+3+1)\n",
    "                    plus_ideal_end.append(proto_start+8)\n",
    "                    plus_ideal_window.append(str(chr9[proto_start+3: proto_start+8]).upper())\n",
    "                    plus_exon_list.append(exon_list[idx1])\n",
    "\n",
    "            #then the minus strand\n",
    "            if k=='C':\n",
    "\n",
    "                if idx+20<=len(sub_chrom):\n",
    "                    minus_count+=1\n",
    "\n",
    "                    minus_pam_loc_end.append(loc_list[idx])\n",
    "                    minus_pam_loc_start.append(loc_list[idx]+1)\n",
    "\n",
    "                    PAM_minus.append(str(chr9[loc_list[idx]:loc_list[idx]+2].reverse_complement()).upper())\n",
    "\n",
    "                    proto_end = loc_list[idx]+2\n",
    "                    proto_start = loc_list[idx] + 22\n",
    "                    proto = str(chr9[proto_end: proto_start].reverse_complement()).upper()\n",
    "\n",
    "                    minus_protospacer.append(proto)\n",
    "                    minus_proto_start.append(proto_start)\n",
    "                    minus_proto_end.append(proto_end+1)\n",
    "\n",
    "                    #and ideal editing window\n",
    "                    minus_ideal_start.append(proto_start-3)\n",
    "                    minus_ideal_end.append(proto_start-8+1)\n",
    "                    minus_ideal_window.append(str(chr9[proto_start-8: proto_start-3].reverse_complement()).upper())\n",
    "                    minus_exon_list.append(exon_list[idx1])\n",
    "\n",
    "    col_names = ['protospacer', 'exon', 'proto_start', 'proto_end', 'PAM', 'PAM_start', 'PAM_end', 'ideal_start', 'ideal_end', 'ideal_window']\n",
    "\n",
    "    cols = [plus_protospacer, plus_exon_list, plus_proto_start,plus_proto_end,PAM_plus,plus_pam_loc_start,plus_pam_loc_end, plus_ideal_start, plus_ideal_end, plus_ideal_window ]\n",
    "    plus_df = pd.DataFrame(dict(zip(col_names, cols)))\n",
    "    plus_df['strand'] = '+'\n",
    "\n",
    "\n",
    "    cols_minus = [minus_protospacer, minus_exon_list, minus_proto_start,minus_proto_end,PAM_minus,minus_pam_loc_start,minus_pam_loc_end, minus_ideal_start, minus_ideal_end, minus_ideal_window ]\n",
    "\n",
    "    minus_df = pd.DataFrame(dict(zip(col_names, cols_minus)))\n",
    "    minus_df['strand'] = '-'\n",
    "    \n",
    "    df = pd.concat((plus_df, minus_df))\n",
    "\n",
    "    #and determine distance to the nearest exon of each guide\n",
    "    distances = []\n",
    "    for k, val in df.iterrows():\n",
    "        s = val['proto_start']\n",
    "        e = val['proto_end']\n",
    "\n",
    "\n",
    "        #make sure it doesn't fall in an exon, and compute distance to the nearest exon (minimum)\n",
    "\n",
    "        distance_to_exon = []\n",
    "\n",
    "        for j in start_end_cds:\n",
    "            start_exon = j[0]\n",
    "            end_exon = j[1]\n",
    "            \n",
    "            assert ((s>=start_exon) & (s<end_exon))!=True, print(val, j)\n",
    "            assert ((e>=start_exon) & (e<end_exon))!=True, print(val, j)\n",
    "\n",
    "            distances1 = [abs(s-start_exon), abs(e-end_exon), abs(e-start_exon), abs(e-end_exon)]\n",
    "\n",
    "            distance_to_exon.append(min(distances1))\n",
    "\n",
    "        \n",
    "        distances.append(min(distance_to_exon))\n",
    "\n",
    "    df['nearest_exon_distance'] = distances\n",
    "\n",
    "\n",
    "    if chrom != 'chrX':\n",
    "        df['chrom'] = int(chrom[3:])\n",
    "    else:\n",
    "        df['chrom'] = chrom[3:]\n",
    "    \n",
    "\n",
    "    df['proto_G+19'] = [f'G{i[1:]}' for i in df['protospacer']]\n",
    "\n",
    "    #and generate the sensors as well\n",
    "    sensor_rc = sensor_maker(df, chr9)\n",
    "    df['sensor_wt'] = sensor_rc\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "intron_holder = []\n",
    "gene_list = []\n",
    "for i, val in fgfr2_3b.iterrows():\n",
    "#for tx in list(cdks['Transcript ID']):\n",
    "    tx = val['Transcript ID']\n",
    "    gene = val['Gene']\n",
    "    gene_list.append(gene)\n",
    "    out = mutagenize_intron(tx, chrom_dict)\n",
    "    out['Gene'] = gene\n",
    "    intron_holder.append(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protospacer</th>\n",
       "      <th>exon</th>\n",
       "      <th>proto_start</th>\n",
       "      <th>proto_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>ideal_start</th>\n",
       "      <th>ideal_end</th>\n",
       "      <th>ideal_window</th>\n",
       "      <th>strand</th>\n",
       "      <th>nearest_exon_distance</th>\n",
       "      <th>chrom</th>\n",
       "      <th>proto_G+19</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCAGGAGAGCAAATGAGTGT</td>\n",
       "      <td>intron1_2</td>\n",
       "      <td>6385955</td>\n",
       "      <td>6385974</td>\n",
       "      <td>TG</td>\n",
       "      <td>6385975</td>\n",
       "      <td>6385976</td>\n",
       "      <td>6385958</td>\n",
       "      <td>6385962</td>\n",
       "      <td>GGAGA</td>\n",
       "      <td>+</td>\n",
       "      <td>39</td>\n",
       "      <td>19</td>\n",
       "      <td>GCAGGAGAGCAAATGAGTGT</td>\n",
       "      <td>AGAGGTAAGACAACACTCATTTGCTCTCCTGACCACCTGTCT</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGTTGTCTTACCTCTTCTCA</td>\n",
       "      <td>intron1_2</td>\n",
       "      <td>6385972</td>\n",
       "      <td>6385991</td>\n",
       "      <td>TG</td>\n",
       "      <td>6385992</td>\n",
       "      <td>6385993</td>\n",
       "      <td>6385975</td>\n",
       "      <td>6385979</td>\n",
       "      <td>TGTCT</td>\n",
       "      <td>+</td>\n",
       "      <td>56</td>\n",
       "      <td>19</td>\n",
       "      <td>GGTTGTCTTACCTCTTCTCA</td>\n",
       "      <td>ATTGTTCAGACATGAGAAGAGGTAAGACAACACTCATTTGCT</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTCTTACCTCTTCTCATGTC</td>\n",
       "      <td>intron1_2</td>\n",
       "      <td>6385976</td>\n",
       "      <td>6385995</td>\n",
       "      <td>TG</td>\n",
       "      <td>6385996</td>\n",
       "      <td>6385997</td>\n",
       "      <td>6385979</td>\n",
       "      <td>6385983</td>\n",
       "      <td>TTACC</td>\n",
       "      <td>+</td>\n",
       "      <td>60</td>\n",
       "      <td>19</td>\n",
       "      <td>GTCTTACCTCTTCTCATGTC</td>\n",
       "      <td>CCATATTGTTCAGACATGAGAAGAGGTAAGACAACACTCATT</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTTCTCATGTCTGAACAATA</td>\n",
       "      <td>intron1_2</td>\n",
       "      <td>6385985</td>\n",
       "      <td>6386004</td>\n",
       "      <td>TG</td>\n",
       "      <td>6386005</td>\n",
       "      <td>6386006</td>\n",
       "      <td>6385988</td>\n",
       "      <td>6385992</td>\n",
       "      <td>CTCAT</td>\n",
       "      <td>+</td>\n",
       "      <td>69</td>\n",
       "      <td>19</td>\n",
       "      <td>GTTCTCATGTCTGAACAATA</td>\n",
       "      <td>GGCTATAAGCCATATTGTTCAGACATGAGAAGAGGTAAGACA</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTCTCATGTCTGAACAATAT</td>\n",
       "      <td>intron1_2</td>\n",
       "      <td>6385986</td>\n",
       "      <td>6386005</td>\n",
       "      <td>GG</td>\n",
       "      <td>6386006</td>\n",
       "      <td>6386007</td>\n",
       "      <td>6385989</td>\n",
       "      <td>6385993</td>\n",
       "      <td>TCATG</td>\n",
       "      <td>+</td>\n",
       "      <td>70</td>\n",
       "      <td>19</td>\n",
       "      <td>GTCTCATGTCTGAACAATAT</td>\n",
       "      <td>GGGCTATAAGCCATATTGTTCAGACATGAGAAGAGGTAAGAC</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>GGATCATCCGTGGGACCTCT</td>\n",
       "      <td>intron8_9</td>\n",
       "      <td>6389683</td>\n",
       "      <td>6389664</td>\n",
       "      <td>AG</td>\n",
       "      <td>6389662</td>\n",
       "      <td>6389661</td>\n",
       "      <td>6389680</td>\n",
       "      <td>6389676</td>\n",
       "      <td>TCATC</td>\n",
       "      <td>-</td>\n",
       "      <td>34</td>\n",
       "      <td>19</td>\n",
       "      <td>GGATCATCCGTGGGACCTCT</td>\n",
       "      <td>GGAGCTGAGCCTAGAGGTCCCACGGATGATCCTCAGCCTTTC</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>AAGGCTGAGGATCATCCGTG</td>\n",
       "      <td>intron8_9</td>\n",
       "      <td>6389691</td>\n",
       "      <td>6389672</td>\n",
       "      <td>GG</td>\n",
       "      <td>6389670</td>\n",
       "      <td>6389669</td>\n",
       "      <td>6389688</td>\n",
       "      <td>6389684</td>\n",
       "      <td>GCTGA</td>\n",
       "      <td>-</td>\n",
       "      <td>26</td>\n",
       "      <td>19</td>\n",
       "      <td>GAGGCTGAGGATCATCCGTG</td>\n",
       "      <td>GCCTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCT</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>AAAGGCTGAGGATCATCCGT</td>\n",
       "      <td>intron8_9</td>\n",
       "      <td>6389692</td>\n",
       "      <td>6389673</td>\n",
       "      <td>GG</td>\n",
       "      <td>6389671</td>\n",
       "      <td>6389670</td>\n",
       "      <td>6389689</td>\n",
       "      <td>6389685</td>\n",
       "      <td>GGCTG</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>GAAGGCTGAGGATCATCCGT</td>\n",
       "      <td>CCTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTT</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>GAAAGGCTGAGGATCATCCG</td>\n",
       "      <td>intron8_9</td>\n",
       "      <td>6389693</td>\n",
       "      <td>6389674</td>\n",
       "      <td>TG</td>\n",
       "      <td>6389672</td>\n",
       "      <td>6389671</td>\n",
       "      <td>6389690</td>\n",
       "      <td>6389686</td>\n",
       "      <td>AGGCT</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>GAAAGGCTGAGGATCATCCG</td>\n",
       "      <td>CTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTTG</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>GTGAAAGGCTGAGGATCATC</td>\n",
       "      <td>intron8_9</td>\n",
       "      <td>6389695</td>\n",
       "      <td>6389676</td>\n",
       "      <td>CG</td>\n",
       "      <td>6389674</td>\n",
       "      <td>6389673</td>\n",
       "      <td>6389692</td>\n",
       "      <td>6389688</td>\n",
       "      <td>AAAGG</td>\n",
       "      <td>-</td>\n",
       "      <td>22</td>\n",
       "      <td>19</td>\n",
       "      <td>GTGAAAGGCTGAGGATCATC</td>\n",
       "      <td>AGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTTGCG</td>\n",
       "      <td>Men1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1242 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              protospacer       exon  proto_start  proto_end PAM  PAM_start  \\\n",
       "0    TCAGGAGAGCAAATGAGTGT  intron1_2      6385955    6385974  TG    6385975   \n",
       "1    TGTTGTCTTACCTCTTCTCA  intron1_2      6385972    6385991  TG    6385992   \n",
       "2    GTCTTACCTCTTCTCATGTC  intron1_2      6385976    6385995  TG    6385996   \n",
       "3    CTTCTCATGTCTGAACAATA  intron1_2      6385985    6386004  TG    6386005   \n",
       "4    TTCTCATGTCTGAACAATAT  intron1_2      6385986    6386005  GG    6386006   \n",
       "..                    ...        ...          ...        ...  ..        ...   \n",
       "533  GGATCATCCGTGGGACCTCT  intron8_9      6389683    6389664  AG    6389662   \n",
       "534  AAGGCTGAGGATCATCCGTG  intron8_9      6389691    6389672  GG    6389670   \n",
       "535  AAAGGCTGAGGATCATCCGT  intron8_9      6389692    6389673  GG    6389671   \n",
       "536  GAAAGGCTGAGGATCATCCG  intron8_9      6389693    6389674  TG    6389672   \n",
       "537  GTGAAAGGCTGAGGATCATC  intron8_9      6389695    6389676  CG    6389674   \n",
       "\n",
       "     PAM_end  ideal_start  ideal_end ideal_window strand  \\\n",
       "0    6385976      6385958    6385962        GGAGA      +   \n",
       "1    6385993      6385975    6385979        TGTCT      +   \n",
       "2    6385997      6385979    6385983        TTACC      +   \n",
       "3    6386006      6385988    6385992        CTCAT      +   \n",
       "4    6386007      6385989    6385993        TCATG      +   \n",
       "..       ...          ...        ...          ...    ...   \n",
       "533  6389661      6389680    6389676        TCATC      -   \n",
       "534  6389669      6389688    6389684        GCTGA      -   \n",
       "535  6389670      6389689    6389685        GGCTG      -   \n",
       "536  6389671      6389690    6389686        AGGCT      -   \n",
       "537  6389673      6389692    6389688        AAAGG      -   \n",
       "\n",
       "     nearest_exon_distance  chrom            proto_G+19  \\\n",
       "0                       39     19  GCAGGAGAGCAAATGAGTGT   \n",
       "1                       56     19  GGTTGTCTTACCTCTTCTCA   \n",
       "2                       60     19  GTCTTACCTCTTCTCATGTC   \n",
       "3                       69     19  GTTCTCATGTCTGAACAATA   \n",
       "4                       70     19  GTCTCATGTCTGAACAATAT   \n",
       "..                     ...    ...                   ...   \n",
       "533                     34     19  GGATCATCCGTGGGACCTCT   \n",
       "534                     26     19  GAGGCTGAGGATCATCCGTG   \n",
       "535                     25     19  GAAGGCTGAGGATCATCCGT   \n",
       "536                     24     19  GAAAGGCTGAGGATCATCCG   \n",
       "537                     22     19  GTGAAAGGCTGAGGATCATC   \n",
       "\n",
       "                                      sensor_wt  Gene  \n",
       "0    AGAGGTAAGACAACACTCATTTGCTCTCCTGACCACCTGTCT  Men1  \n",
       "1    ATTGTTCAGACATGAGAAGAGGTAAGACAACACTCATTTGCT  Men1  \n",
       "2    CCATATTGTTCAGACATGAGAAGAGGTAAGACAACACTCATT  Men1  \n",
       "3    GGCTATAAGCCATATTGTTCAGACATGAGAAGAGGTAAGACA  Men1  \n",
       "4    GGGCTATAAGCCATATTGTTCAGACATGAGAAGAGGTAAGAC  Men1  \n",
       "..                                          ...   ...  \n",
       "533  GGAGCTGAGCCTAGAGGTCCCACGGATGATCCTCAGCCTTTC  Men1  \n",
       "534  GCCTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCT  Men1  \n",
       "535  CCTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTT  Men1  \n",
       "536  CTAGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTTG  Men1  \n",
       "537  AGAGGTCCCACGGATGATCCTCAGCCTTTCACTTGGCTTGCG  Men1  \n",
       "\n",
       "[1242 rows x 16 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_dict = dict(zip(gene_list, intron_holder))\n",
    "intron_dict['Men1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_exon_dist_cutoff = 30\n",
    "num_guides = 2\n",
    "\n",
    "holder = []\n",
    "for i in intron_dict.keys():\n",
    "\n",
    "    cdk9_introns = intron_dict[i]\n",
    "    intron_names = np.unique(cdk9_introns['exon'])\n",
    "\n",
    "    cdk9_introns = cdk9_introns[cdk9_introns['nearest_exon_distance']>=min_exon_dist_cutoff]\n",
    "\n",
    "    #then pre-filter out polyT sequences and RE sites\n",
    "    RE_sites_polyT = ['CGTCTC', 'GAATTC', 'GAGACG', 'TTTT']\n",
    "\n",
    "    filt = []\n",
    "    for i, val in cdk9_introns.iterrows():\n",
    "        p= val['protospacer']\n",
    "        s = val['sensor_wt']\n",
    "        \n",
    "        c = 0\n",
    "        for k in RE_sites_polyT:\n",
    "            if k in p:\n",
    "                c+=1\n",
    "            if k in s:\n",
    "                c+=1\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        if c>0:\n",
    "            filt.append(True)\n",
    "        else:\n",
    "            filt.append(False)\n",
    "\n",
    "    \n",
    "    cdk9_introns['polyT_or_RE_site'] = filt\n",
    "    cdk9_introns = cdk9_introns[cdk9_introns['polyT_or_RE_site']==False].reset_index()\n",
    "\n",
    "    #and then randomly picking 4 guides per intron\n",
    "    for k in intron_names:\n",
    "        subset = cdk9_introns[cdk9_introns['exon']==k].reset_index()\n",
    "\n",
    "        a = list(range(len(subset)))\n",
    "        #print(min(num_guides, len(a)))\n",
    "        choices = np.random.choice(a, size=min(num_guides, len(a)), replace=False)\n",
    "        \n",
    "        holder.append(subset.iloc[choices])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['Kdm6a', 'Kmt2a', 'Kmt2b', 'Kmt2c', 'Kmt2d', 'Men1'], dtype=object),\n",
       " array([ 54,  69,  66, 113, 101,  16]))"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_guides_select = pd.concat(holder)\n",
    "intron_guides_select = intron_guides_select.drop(columns = ['level_0', 'index'])\n",
    "intron_guides_select = intron_guides_select.reset_index().drop(columns=['index'])\n",
    "intron_guides_select\n",
    "\n",
    "#and saving i\n",
    "#intron_guides_select.to_csv('intron_guides_selected/intron_guides_selected.csv', index=False)\n",
    "np.unique(intron_guides_select['Gene'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Men1 : 16/16\n",
      "Kdm6a : 54/56\n",
      "Kmt2a : 69/70\n",
      "Kmt2b : 66/72\n",
      "Kmt2c : 113/116\n",
      "Kmt2d : 101/106\n"
     ]
    }
   ],
   "source": [
    "u, c = np.unique(intron_guides_select['Gene'], return_counts=True)\n",
    "dict_i = dict(zip(u, c))\n",
    "\n",
    "for i in intron_dict.keys():\n",
    "\n",
    "    cdk9_introns = intron_dict[i]\n",
    "    intron_names = np.unique(cdk9_introns['exon'])\n",
    "\n",
    "    print(f'{i} : {dict_i[i]}/{len(intron_names)*num_guides}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combining, adding BCs, and generating library\n",
    "\n",
    "- Filtering for RE sites and polyT sites as well\n",
    "- And creating a new gRNA_id for each (should include the gene name)\n",
    "- Add 5% non-targeting controls\n",
    "    - Want these to overlap for different subpools...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Men1', 'Kdm6a', 'Kmt2a', 'Kmt2b', 'Kmt2c', 'Kmt2d']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = list(fgfr2_3b['Gene'])\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>proto_G+19</th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon</th>\n",
       "      <th>proto_start</th>\n",
       "      <th>proto_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>PAM_end</th>\n",
       "      <th>ideal_start</th>\n",
       "      <th>ideal_end</th>\n",
       "      <th>ideal_window</th>\n",
       "      <th>strand</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>CDS_hit</th>\n",
       "      <th>ABE_amenable</th>\n",
       "      <th>CBE_amenable</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gRNA_1</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GCCGCCATGGGGCTGAAGGC</td>\n",
       "      <td>GCCGCCATGGGGCTGAAGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385485</td>\n",
       "      <td>6385504</td>\n",
       "      <td>CG</td>\n",
       "      <td>6385505</td>\n",
       "      <td>6385506</td>\n",
       "      <td>6385488</td>\n",
       "      <td>6385492</td>\n",
       "      <td>GCCAT</td>\n",
       "      <td>+</td>\n",
       "      <td>GTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGGGCGGTGGC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gRNA_2</td>\n",
       "      <td>Men1</td>\n",
       "      <td>CATGGGGCTGAAGGCCGCCC</td>\n",
       "      <td>GATGGGGCTGAAGGCCGCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385490</td>\n",
       "      <td>6385509</td>\n",
       "      <td>AG</td>\n",
       "      <td>6385510</td>\n",
       "      <td>6385511</td>\n",
       "      <td>6385493</td>\n",
       "      <td>6385497</td>\n",
       "      <td>GGGGC</td>\n",
       "      <td>+</td>\n",
       "      <td>ACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGGGCG</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gRNA_3</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GGGGCTGAAGGCCGCCCAGA</td>\n",
       "      <td>GGGGCTGAAGGCCGCCCAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385493</td>\n",
       "      <td>6385512</td>\n",
       "      <td>AG</td>\n",
       "      <td>6385513</td>\n",
       "      <td>6385514</td>\n",
       "      <td>6385496</td>\n",
       "      <td>6385500</td>\n",
       "      <td>GCTGA</td>\n",
       "      <td>+</td>\n",
       "      <td>GGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGG</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gRNA_4</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GCTGAAGGCCGCCCAGAAGA</td>\n",
       "      <td>GCTGAAGGCCGCCCAGAAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385496</td>\n",
       "      <td>6385515</td>\n",
       "      <td>CG</td>\n",
       "      <td>6385516</td>\n",
       "      <td>6385517</td>\n",
       "      <td>6385499</td>\n",
       "      <td>6385503</td>\n",
       "      <td>GAAGG</td>\n",
       "      <td>+</td>\n",
       "      <td>GAGGGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGG</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gRNA_5</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GAAGGCCGCCCAGAAGACGC</td>\n",
       "      <td>GAAGGCCGCCCAGAAGACGC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385499</td>\n",
       "      <td>6385518</td>\n",
       "      <td>TG</td>\n",
       "      <td>6385519</td>\n",
       "      <td>6385520</td>\n",
       "      <td>6385502</td>\n",
       "      <td>6385506</td>\n",
       "      <td>GGCCG</td>\n",
       "      <td>+</td>\n",
       "      <td>GCAGAGGGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGG</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4678</th>\n",
       "      <td>gRNA_10239</td>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>AGCCAGAAGCCGCCTGCTGA</td>\n",
       "      <td>GGCCAGAAGCCGCCTGCTGA</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>98764595</td>\n",
       "      <td>98764576</td>\n",
       "      <td>AG</td>\n",
       "      <td>98764574</td>\n",
       "      <td>98764573</td>\n",
       "      <td>98764592</td>\n",
       "      <td>98764588</td>\n",
       "      <td>CAGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>GAATCTTTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCT</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4679</th>\n",
       "      <td>gRNA_10240</td>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>GACAGCCAGAAGCCGCCTGC</td>\n",
       "      <td>GACAGCCAGAAGCCGCCTGC</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>98764598</td>\n",
       "      <td>98764579</td>\n",
       "      <td>TG</td>\n",
       "      <td>98764577</td>\n",
       "      <td>98764576</td>\n",
       "      <td>98764595</td>\n",
       "      <td>98764591</td>\n",
       "      <td>AGCCA</td>\n",
       "      <td>-</td>\n",
       "      <td>TCTTTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4680</th>\n",
       "      <td>gRNA_10241</td>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>ATGGACAGCCAGAAGCCGCC</td>\n",
       "      <td>GTGGACAGCCAGAAGCCGCC</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>98764601</td>\n",
       "      <td>98764582</td>\n",
       "      <td>TG</td>\n",
       "      <td>98764580</td>\n",
       "      <td>98764579</td>\n",
       "      <td>98764598</td>\n",
       "      <td>98764594</td>\n",
       "      <td>GACAG</td>\n",
       "      <td>-</td>\n",
       "      <td>TTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4681</th>\n",
       "      <td>gRNA_10242</td>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>AGGGATGGACAGCCAGAAGC</td>\n",
       "      <td>GGGGATGGACAGCCAGAAGC</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>98764605</td>\n",
       "      <td>98764586</td>\n",
       "      <td>CG</td>\n",
       "      <td>98764584</td>\n",
       "      <td>98764583</td>\n",
       "      <td>98764602</td>\n",
       "      <td>98764598</td>\n",
       "      <td>GATGG</td>\n",
       "      <td>-</td>\n",
       "      <td>CTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCCGACT</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4682</th>\n",
       "      <td>gRNA_10243</td>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>GAGAGGGATGGACAGCCAGA</td>\n",
       "      <td>GAGAGGGATGGACAGCCAGA</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>98764608</td>\n",
       "      <td>98764589</td>\n",
       "      <td>AG</td>\n",
       "      <td>98764587</td>\n",
       "      <td>98764586</td>\n",
       "      <td>98764605</td>\n",
       "      <td>98764601</td>\n",
       "      <td>AGGGA</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCCGACTGGG</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>targeting</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31599 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gRNA_id   Gene           protospacer            proto_G+19 chrom  \\\n",
       "0         gRNA_1   Men1  GCCGCCATGGGGCTGAAGGC  GCCGCCATGGGGCTGAAGGC    19   \n",
       "1         gRNA_2   Men1  CATGGGGCTGAAGGCCGCCC  GATGGGGCTGAAGGCCGCCC    19   \n",
       "2         gRNA_3   Men1  GGGGCTGAAGGCCGCCCAGA  GGGGCTGAAGGCCGCCCAGA    19   \n",
       "3         gRNA_4   Men1  GCTGAAGGCCGCCCAGAAGA  GCTGAAGGCCGCCCAGAAGA    19   \n",
       "4         gRNA_5   Men1  GAAGGCCGCCCAGAAGACGC  GAAGGCCGCCCAGAAGACGC    19   \n",
       "...          ...    ...                   ...                   ...   ...   \n",
       "4678  gRNA_10239  Kmt2d  AGCCAGAAGCCGCCTGCTGA  GGCCAGAAGCCGCCTGCTGA    15   \n",
       "4679  gRNA_10240  Kmt2d  GACAGCCAGAAGCCGCCTGC  GACAGCCAGAAGCCGCCTGC    15   \n",
       "4680  gRNA_10241  Kmt2d  ATGGACAGCCAGAAGCCGCC  GTGGACAGCCAGAAGCCGCC    15   \n",
       "4681  gRNA_10242  Kmt2d  AGGGATGGACAGCCAGAAGC  GGGGATGGACAGCCAGAAGC    15   \n",
       "4682  gRNA_10243  Kmt2d  GAGAGGGATGGACAGCCAGA  GAGAGGGATGGACAGCCAGA    15   \n",
       "\n",
       "      exon  proto_start  proto_end PAM  PAM_start   PAM_end  ideal_start  \\\n",
       "0        1      6385485    6385504  CG    6385505   6385506      6385488   \n",
       "1        1      6385490    6385509  AG    6385510   6385511      6385493   \n",
       "2        1      6385493    6385512  AG    6385513   6385514      6385496   \n",
       "3        1      6385496    6385515  CG    6385516   6385517      6385499   \n",
       "4        1      6385499    6385518  TG    6385519   6385520      6385502   \n",
       "...    ...          ...        ...  ..        ...       ...          ...   \n",
       "4678     1     98764595   98764576  AG   98764574  98764573     98764592   \n",
       "4679     1     98764598   98764579  TG   98764577  98764576     98764595   \n",
       "4680     1     98764601   98764582  TG   98764580  98764579     98764598   \n",
       "4681     1     98764605   98764586  CG   98764584  98764583     98764602   \n",
       "4682     1     98764608   98764589  AG   98764587  98764586     98764605   \n",
       "\n",
       "      ideal_end ideal_window strand  \\\n",
       "0       6385492        GCCAT      +   \n",
       "1       6385497        GGGGC      +   \n",
       "2       6385500        GCTGA      +   \n",
       "3       6385503        GAAGG      +   \n",
       "4       6385506        GGCCG      +   \n",
       "...         ...          ...    ...   \n",
       "4678   98764588        CAGAA      -   \n",
       "4679   98764591        AGCCA      -   \n",
       "4680   98764594        GACAG      -   \n",
       "4681   98764598        GATGG      -   \n",
       "4682   98764601        AGGGA      -   \n",
       "\n",
       "                                       sensor_wt  CDS_hit  ABE_amenable  \\\n",
       "0     GTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGGGCGGTGGC     True          True   \n",
       "1     ACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGGGCG     True         False   \n",
       "2     GGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGGCGG     True          True   \n",
       "3     GAGGGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGGCGG     True          True   \n",
       "4     GCAGAGGGAACAGCGTCTTCTGGGCGGCCTTCAGCCCCATGG     True         False   \n",
       "...                                          ...      ...           ...   \n",
       "4678  GAATCTTTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCT     True          True   \n",
       "4679  TCTTTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTC     True          True   \n",
       "4680  TTATCTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCC     True          True   \n",
       "4681  CTTCAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCCGACT     True          True   \n",
       "4682  CAGCAGGCGGCTTCTGGCTGTCCATCCCTCTCTCCGACTGGG     True          True   \n",
       "\n",
       "      CBE_amenable classification  \n",
       "0             True      targeting  \n",
       "1             True      targeting  \n",
       "2             True      targeting  \n",
       "3            False      targeting  \n",
       "4             True      targeting  \n",
       "...            ...            ...  \n",
       "4678          True      targeting  \n",
       "4679          True      targeting  \n",
       "4680          True      targeting  \n",
       "4681         False      targeting  \n",
       "4682         False      targeting  \n",
       "\n",
       "[31599 rows x 20 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#first combine the guide\n",
    "genes = list(fgfr2_3b['Gene'])\n",
    "\n",
    "guide_holder = []\n",
    "for i in genes:\n",
    "    d = guide_dict[i]\n",
    "    d['Gene'] = i\n",
    "    guide_holder.append(d)\n",
    "\n",
    "cols = ['gRNA_id', 'Gene','protospacer', 'proto_G+19', 'chrom', 'exon', 'proto_start',\n",
    "       'proto_end', 'PAM', 'PAM_start', 'PAM_end', 'ideal_start', 'ideal_end',\n",
    "       'ideal_window', 'strand', 'sensor_wt', 'CDS_hit', 'ABE_amenable',\n",
    "       'CBE_amenable']\n",
    "\n",
    "combined_guides = pd.concat(guide_holder)[cols]\n",
    "combined_guides['classification'] = 'targeting'\n",
    "combined_guides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Men1', 'Kdm6a', 'Kmt2a', 'Kmt2b', 'Kmt2c', 'Kmt2d']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#nontargeting = pd.read_csv('non_targeting_guides.csv').rename(columns = {'Protospacer':'protospacer', 'sensor_seq_random':'sensor_wt'}).drop(columns = ['polyT_or_RE_site'])\n",
    "nontargeting = nontarget.rename(columns = {'Protospacer':'protospacer', 'sensor_seq_random':'sensor_wt'}).drop(columns = ['polyT_or_RE_site'])\n",
    "\n",
    "intron_guides_select['classification'] = 'intron'\n",
    "\n",
    "#combining exon and intron-targeting guides\n",
    "targeting = pd.concat((combined_guides, intron_guides_select))\n",
    "\n",
    "targeting.loc[targeting['Gene']==genes[0], 'Pool'] = 'F2-R1'\n",
    "targeting.loc[targeting['Gene']==genes[1], 'Pool'] = 'F2-R2'\n",
    "targeting.loc[targeting['Gene']==genes[2], 'Pool'] = 'F2-R3'\n",
    "targeting.loc[targeting['Gene']==genes[3], 'Pool'] = 'F2-R4'\n",
    "targeting.loc[targeting['Gene']==genes[4], 'Pool'] = 'F2-R5'\n",
    "targeting.loc[targeting['Gene']==genes[5], 'Pool'] = 'F2-R6'\n",
    "\n",
    "targeting = targeting.drop(columns = ['polyT_or_RE_site'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#determining number of non-targeting guides to include for each\n",
    "fraction_nt = .03\n",
    "\n",
    "size1 = int(len(targeting[targeting['Pool']=='F2-R1'])*fraction_nt)\n",
    "size2 = int(len(targeting[targeting['Pool']=='F2-R2'])*fraction_nt)\n",
    "size3 = int(len(targeting[targeting['Pool']=='F2-R3'])*fraction_nt)\n",
    "size4 = int(len(targeting[targeting['Pool']=='F2-R4'])*fraction_nt)\n",
    "size5 = int(len(targeting[targeting['Pool']=='F2-R5'])*fraction_nt)\n",
    "size6 = int(len(targeting[targeting['Pool']=='F2-R6'])*fraction_nt)\n",
    "\n",
    "nt_subpool1 = nontargeting[:size1]\n",
    "nt_subpool2 = nontargeting[:size2]\n",
    "nt_subpool3 = nontargeting[:size3]\n",
    "nt_subpool4 = nontargeting[:size4]\n",
    "nt_subpool5 = nontargeting[:size5]\n",
    "nt_subpool6 = nontargeting[:size6]\n",
    "\n",
    "nt_subpool1['Pool'] = 'F2-R1'\n",
    "nt_subpool2['Pool'] = 'F2-R2'\n",
    "nt_subpool3['Pool'] = 'F2-R3'\n",
    "nt_subpool4['Pool'] = 'F2-R4'\n",
    "nt_subpool5['Pool'] = 'F2-R5'\n",
    "nt_subpool6['Pool'] = 'F2-R6'\n",
    "\n",
    "#and then adding these to the rest of the guides\n",
    "combined_library = pd.concat([targeting, nt_subpool1, nt_subpool2, nt_subpool3, nt_subpool4, nt_subpool5, nt_subpool6])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'intron': 16, 'non-targeting control': 34, 'targeting': 1129}\n",
      "{'intron': 54, 'non-targeting control': 58, 'targeting': 1903}\n",
      "{'intron': 69, 'non-targeting control': 191, 'targeting': 6320}\n",
      "{'intron': 66, 'non-targeting control': 152, 'targeting': 5029}\n",
      "{'intron': 113, 'non-targeting control': 212, 'targeting': 6975}\n",
      "{'intron': 101, 'non-targeting control': 310, 'targeting': 10243}\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,7):\n",
    "    pool = f'F2-R{i}'\n",
    "    subset = combined_library[combined_library['Pool']==pool]\n",
    "    u, c = np.unique(subset['classification'], return_counts=True)\n",
    "    print(dict(zip(u,c)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adding hamming barcodes\n",
    "hamming = pd.read_csv('Hamming_BC_fil.csv')\n",
    "\n",
    "#modified so there's no overlap in barcode\n",
    "hamming = hamming[40000:]\n",
    "\n",
    "#first filter out RE sites EcoRI and Esp3I\n",
    "RE_sites = ['CGTCTC', 'GAATTC', 'GAGACG']\n",
    "\n",
    "contains_RE = []\n",
    "for i in list(hamming['Hammingbarcode']):\n",
    "    c=0\n",
    "    for k in RE_sites_polyT:\n",
    "        if k in i:\n",
    "            c+=1\n",
    "    \n",
    "    if c>0:\n",
    "        contains_RE.append(True)\n",
    "    elif c==0:\n",
    "        contains_RE.append(False)\n",
    "\n",
    "hamming['contains_RE'] = contains_RE\n",
    "\n",
    "ham = hamming[hamming['contains_RE']==False]\n",
    "#decreasing the size to 13 to keep oligo length at 250\n",
    "ham_bc = [i[:13] for i in list(ham['Hammingbarcode'])]\n",
    "\n",
    "combined_library['Hamming_BC'] = ham_bc[:len(combined_library)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#------oligo generation tool\n",
    "#----taken/modified from PEGG.base\n",
    "def base_oligo_generator(peg_df, five_prime_adapter = 'AGCGTACACGTCTCACACC',three_prime_adapter = 'GAATTCTAGATCCGGTCGTCAAC',\n",
    "                gRNA_scaff = 'GTTTAAGAGCTATGCTGGAAACAGCATAGCAAGTTTAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGC'):\n",
    "\n",
    "    \"\"\"\n",
    "    A tool for automatically generating oligos from the output of run().\n",
    "    Returns input dataframe with new columns containing gRNA oligo with or without sensor.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    peg_df\n",
    "        *type = pd.DataFrame*\n",
    "        \n",
    "        A dataframe containing the gRNAs for the selected input mutations. Generated by run_base() or gRNA_generator()\n",
    "\n",
    "    five_prime_adapter\n",
    "        *type = str*\n",
    "        \n",
    "        5' Prime Adapter. The automatically provided 5' adapter contains an Esp3I (BsmBI) site. Can be swapped with \n",
    "        whatever input string user wants.\n",
    "    \n",
    "    three_prime_adapter\n",
    "        *type = str*\n",
    "        \n",
    "        5' Prime Adapter. The automatically provided 5' adapter contains an Esp3I (BsmBI) site. Can be swapped with \n",
    "        whatever input string user wants.\n",
    "        \n",
    "    gRNA_scaff\n",
    "        *type = str*\n",
    "        \n",
    "        gRNA scaffold region. Automatically set to a functional gRNA scaffold. Can be swapped with \n",
    "        whatever input string user wants.\n",
    "    \n",
    "    \"\"\"\n",
    "        \n",
    "        \n",
    "    u6_term = 'TTTTTTT'\n",
    "\n",
    "    base_oligos = []\n",
    "\n",
    "    for i, val in peg_df.iterrows():\n",
    "        proto = val['proto_G+19']\n",
    "        #extension = val['RTT_PBS']\n",
    "        sensor = val[\"sensor_wt\"]\n",
    "        bc = val['Hamming_BC']\n",
    "\n",
    "        pool = val['Pool']\n",
    "\n",
    "        f_identity = pool.split('-')[0]\n",
    "        r_identity = pool.split('-')[1]\n",
    "\n",
    "        f_dict = dict(zip(['F1','F2', 'F3', 'F4', 'F5', 'F6'],\n",
    "                          ['AGGCACTTGCTCGTACGACG',\n",
    "                           'GTGTAACCCGTAGGGCACCT',\n",
    "                           'CAGCGCCAATGGGCTTTCGA',\n",
    "                           'CTACAGGTACCGGTCCTGAG',\n",
    "                           'CATGTTGCCCTGAGGCACAG',\n",
    "                           'GGTCGTCGCATCACAATGCG']))\n",
    "        \n",
    "        r_dict = dict(zip(['R1','R2','R3','R4','R5','R6'],\n",
    "                          ['TTAAGGTGCCGGGCCCACAT',\n",
    "                           'GTCGAAGGACTGCTCTCGAC',\n",
    "                           'CGACAGGCTCTTAAGCGGCT',\n",
    "                           'CGGATCGTCACGCTAGGTAC',\n",
    "                           'AGCCTTTCGGGACCTAACGG',\n",
    "                           'CGTCACATTGGCGCTCGAGA']))\n",
    "\n",
    "\n",
    "        F = f_dict[f_identity]\n",
    "        R = r_dict[r_identity]\n",
    "\n",
    "        gRNA_full = F + five_prime_adapter + proto + gRNA_scaff + u6_term + sensor + bc + three_prime_adapter + R\n",
    "        base_oligos.append(gRNA_full)\n",
    "\n",
    "    \n",
    "    peg_df['gRNA_oligo'] = base_oligos\n",
    "    \n",
    "    return peg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_library = base_oligo_generator(combined_library)\n",
    "combined_library = combined_library.reset_index().drop(columns='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>polyT_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Men1</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Kdm6a</td>\n",
       "      <td>149.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kmt2a</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Kmt2b</td>\n",
       "      <td>101.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Kmt2c</td>\n",
       "      <td>428.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Kmt2d</td>\n",
       "      <td>212.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene  polyT_count\n",
       "0   Men1         13.0\n",
       "1  Kdm6a        149.0\n",
       "2  Kmt2a        352.0\n",
       "3  Kmt2b        101.0\n",
       "4  Kmt2c        428.0\n",
       "5  Kmt2d        212.0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#counting polyT sequence (not filtering these though)\n",
    "\n",
    "#checking for presence of polyT sequence in proto_G+19\n",
    "polyT = pd.DataFrame(dict(zip(['Gene', 'polyT_count'], [list(fgfr2_3b['Gene']), np.zeros(6)])))\n",
    "for i, val in combined_library.iterrows():\n",
    "    g = val['Gene']\n",
    "\n",
    "    if 'TTTT' in val['proto_G+19']:\n",
    "        polyT.loc[polyT['Gene']==g, 'polyT_count']+=1\n",
    "        combined_library.loc[i, 'polyT'] = True\n",
    "    else:\n",
    "        combined_library.loc[i, 'polyT'] = False\n",
    "\n",
    "\n",
    "polyT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filtering for RE sites (making sure that there are only 2)\n",
    "\n",
    "RE_sites= ['CGTCTC', 'GAATTC', 'GAGACG']\n",
    "num_RE_sites = []\n",
    "\n",
    "count = 0\n",
    "\n",
    "for oligo in list(combined_library['gRNA_oligo']):\n",
    "    total = 0\n",
    "    for j in RE_sites:\n",
    "        total += oligo.count(j)\n",
    "\n",
    "    if total>2:\n",
    "        count+=1\n",
    "    num_RE_sites.append(total)\n",
    "\n",
    "combined_library['RE_sites'] = num_RE_sites\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Kdm6a': 1873, 'Kmt2a': 6185, 'Kmt2b': 4917, 'Kmt2c': 6749, 'Kmt2d': 10089, 'Men1': 1105}\n",
      "{'Kdm6a': 1903, 'Kmt2a': 6320, 'Kmt2b': 5029, 'Kmt2c': 6975, 'Kmt2d': 10243, 'Men1': 1129}\n"
     ]
    }
   ],
   "source": [
    "combined_library_filtered = combined_library[combined_library['RE_sites']==2].reset_index()\n",
    "combined_library_filtered = combined_library_filtered.rename(columns = {'gRNA_id':'gRNA_id_OLD'})\n",
    "\n",
    "t2 = combined_library_filtered[combined_library_filtered['classification']=='targeting']\n",
    "u, c= np.unique(t2['Gene'], return_counts=True)\n",
    "print(dict(zip(u,c)))\n",
    "t3 = combined_library[combined_library['classification']=='targeting']\n",
    "u, c= np.unique(t3['Gene'], return_counts=True)\n",
    "print(dict(zip(u,c)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_library_filtered\n",
    "\n",
    "\n",
    "type_dict = {'intron':'intron', 'non-targeting control':'nt', 'targeting':'targ'}\n",
    "#adding gRNA_ids to this\n",
    "ids = []\n",
    "for i, val in combined_library_filtered.iterrows():\n",
    "    g = val['Gene']\n",
    "    class1 = val['classification']\n",
    "    class2 = type_dict[class1]\n",
    "\n",
    "    if class2 == 'nt':\n",
    "        name = f'gRNA_mouse_{class2}_{i+1}'\n",
    "    else:\n",
    "        name = f'gRNA_mouse_{g}_{class2}_{i+1}'\n",
    "    \n",
    "    ids.append(name)\n",
    "\n",
    "combined_library_filtered['gRNA_id'] = ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = ['gRNA_id', 'Gene', 'protospacer', 'proto_G+19', 'chrom', 'exon',\n",
    "       'proto_start', 'proto_end', 'PAM', 'PAM_start', 'PAM_end',\n",
    "       'ideal_start', 'ideal_end', 'ideal_window', 'strand', 'sensor_wt',\n",
    "       'CDS_hit', 'ABE_amenable', 'CBE_amenable', 'classification',\n",
    "       'nearest_exon_distance', 'Pool', 'Hamming_BC', 'gRNA_oligo', 'polyT', 'RE_sites', 'gRNA_id_OLD',]\n",
    "\n",
    "combined_library_filtered = combined_library_filtered[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>protospacer</th>\n",
       "      <th>proto_G+19</th>\n",
       "      <th>chrom</th>\n",
       "      <th>exon</th>\n",
       "      <th>proto_start</th>\n",
       "      <th>proto_end</th>\n",
       "      <th>PAM</th>\n",
       "      <th>PAM_start</th>\n",
       "      <th>...</th>\n",
       "      <th>ABE_amenable</th>\n",
       "      <th>CBE_amenable</th>\n",
       "      <th>classification</th>\n",
       "      <th>nearest_exon_distance</th>\n",
       "      <th>Pool</th>\n",
       "      <th>Hamming_BC</th>\n",
       "      <th>gRNA_oligo</th>\n",
       "      <th>polyT</th>\n",
       "      <th>RE_sites</th>\n",
       "      <th>gRNA_id_OLD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gRNA_mouse_Men1_targ_1</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GCCGCCATGGGGCTGAAGGC</td>\n",
       "      <td>GCCGCCATGGGGCTGAAGGC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385485.0</td>\n",
       "      <td>6385504.0</td>\n",
       "      <td>CG</td>\n",
       "      <td>6385505.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R1</td>\n",
       "      <td>TTATACACAGCTA</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCCGCCA...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>gRNA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gRNA_mouse_Men1_targ_2</td>\n",
       "      <td>Men1</td>\n",
       "      <td>CATGGGGCTGAAGGCCGCCC</td>\n",
       "      <td>GATGGGGCTGAAGGCCGCCC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385490.0</td>\n",
       "      <td>6385509.0</td>\n",
       "      <td>AG</td>\n",
       "      <td>6385510.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R1</td>\n",
       "      <td>ATGGTGCTCCTGG</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGATGGGG...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>gRNA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gRNA_mouse_Men1_targ_3</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GGGGCTGAAGGCCGCCCAGA</td>\n",
       "      <td>GGGGCTGAAGGCCGCCCAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385493.0</td>\n",
       "      <td>6385512.0</td>\n",
       "      <td>AG</td>\n",
       "      <td>6385513.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R1</td>\n",
       "      <td>TACGACGAGCGAG</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGGGGCTG...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>gRNA_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gRNA_mouse_Men1_targ_4</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GCTGAAGGCCGCCCAGAAGA</td>\n",
       "      <td>GCTGAAGGCCGCCCAGAAGA</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385496.0</td>\n",
       "      <td>6385515.0</td>\n",
       "      <td>CG</td>\n",
       "      <td>6385516.0</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>targeting</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R1</td>\n",
       "      <td>AGCACGGAAGTGC</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCTGAAG...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>gRNA_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gRNA_mouse_Men1_targ_5</td>\n",
       "      <td>Men1</td>\n",
       "      <td>GAAGGCCGCCCAGAAGACGC</td>\n",
       "      <td>GAAGGCCGCCCAGAAGACGC</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>6385499.0</td>\n",
       "      <td>6385518.0</td>\n",
       "      <td>TG</td>\n",
       "      <td>6385519.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>targeting</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R1</td>\n",
       "      <td>TGGGCCCGCCCAT</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGAAGGCC...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>gRNA_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32283</th>\n",
       "      <td>gRNA_mouse_nt_32284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATTGCGCAATCCTTAGGATA</td>\n",
       "      <td>GTTGCGCAATCCTTAGGATA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R6</td>\n",
       "      <td>CCTGTAATCGTTC</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGTTGCGC...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32284</th>\n",
       "      <td>gRNA_mouse_nt_32285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGAAGCGGGACCGTGTCTCAC</td>\n",
       "      <td>GAAGCGGGACCGTGTCTCAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R6</td>\n",
       "      <td>AGAAACTTACAGC</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGAAGCGG...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32285</th>\n",
       "      <td>gRNA_mouse_nt_32286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGTGTATGATGCTTCGACTTA</td>\n",
       "      <td>GTGTATGATGCTTCGACTTA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R6</td>\n",
       "      <td>AACAAGGTAACGA</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGTGTATG...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32286</th>\n",
       "      <td>gRNA_mouse_nt_32287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCCTACTCCCGTGTGTTATCC</td>\n",
       "      <td>GCTACTCCCGTGTGTTATCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R6</td>\n",
       "      <td>TGTGGGATATAAA</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCTACTC...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32287</th>\n",
       "      <td>gRNA_mouse_nt_32288</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCGCACGACCATTGCTGCTGC</td>\n",
       "      <td>GGCACGACCATTGCTGCTGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-targeting control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F2-R6</td>\n",
       "      <td>TTTGACCTAGTTG</td>\n",
       "      <td>GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGGCACGA...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32288 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      gRNA_id  Gene            protospacer  \\\n",
       "0      gRNA_mouse_Men1_targ_1  Men1   GCCGCCATGGGGCTGAAGGC   \n",
       "1      gRNA_mouse_Men1_targ_2  Men1   CATGGGGCTGAAGGCCGCCC   \n",
       "2      gRNA_mouse_Men1_targ_3  Men1   GGGGCTGAAGGCCGCCCAGA   \n",
       "3      gRNA_mouse_Men1_targ_4  Men1   GCTGAAGGCCGCCCAGAAGA   \n",
       "4      gRNA_mouse_Men1_targ_5  Men1   GAAGGCCGCCCAGAAGACGC   \n",
       "...                       ...   ...                    ...   \n",
       "32283     gRNA_mouse_nt_32284   NaN  GATTGCGCAATCCTTAGGATA   \n",
       "32284     gRNA_mouse_nt_32285   NaN  GGAAGCGGGACCGTGTCTCAC   \n",
       "32285     gRNA_mouse_nt_32286   NaN  GGTGTATGATGCTTCGACTTA   \n",
       "32286     gRNA_mouse_nt_32287   NaN  GCCTACTCCCGTGTGTTATCC   \n",
       "32287     gRNA_mouse_nt_32288   NaN  GCGCACGACCATTGCTGCTGC   \n",
       "\n",
       "                 proto_G+19 chrom exon  proto_start  proto_end  PAM  \\\n",
       "0      GCCGCCATGGGGCTGAAGGC    19    1    6385485.0  6385504.0   CG   \n",
       "1      GATGGGGCTGAAGGCCGCCC    19    1    6385490.0  6385509.0   AG   \n",
       "2      GGGGCTGAAGGCCGCCCAGA    19    1    6385493.0  6385512.0   AG   \n",
       "3      GCTGAAGGCCGCCCAGAAGA    19    1    6385496.0  6385515.0   CG   \n",
       "4      GAAGGCCGCCCAGAAGACGC    19    1    6385499.0  6385518.0   TG   \n",
       "...                     ...   ...  ...          ...        ...  ...   \n",
       "32283  GTTGCGCAATCCTTAGGATA   NaN  NaN          NaN        NaN  NaN   \n",
       "32284  GAAGCGGGACCGTGTCTCAC   NaN  NaN          NaN        NaN  NaN   \n",
       "32285  GTGTATGATGCTTCGACTTA   NaN  NaN          NaN        NaN  NaN   \n",
       "32286  GCTACTCCCGTGTGTTATCC   NaN  NaN          NaN        NaN  NaN   \n",
       "32287  GGCACGACCATTGCTGCTGC   NaN  NaN          NaN        NaN  NaN   \n",
       "\n",
       "       PAM_start  ...  ABE_amenable  CBE_amenable         classification  \\\n",
       "0      6385505.0  ...          True          True              targeting   \n",
       "1      6385510.0  ...         False          True              targeting   \n",
       "2      6385513.0  ...          True          True              targeting   \n",
       "3      6385516.0  ...          True         False              targeting   \n",
       "4      6385519.0  ...         False          True              targeting   \n",
       "...          ...  ...           ...           ...                    ...   \n",
       "32283        NaN  ...           NaN           NaN  non-targeting control   \n",
       "32284        NaN  ...           NaN           NaN  non-targeting control   \n",
       "32285        NaN  ...           NaN           NaN  non-targeting control   \n",
       "32286        NaN  ...           NaN           NaN  non-targeting control   \n",
       "32287        NaN  ...           NaN           NaN  non-targeting control   \n",
       "\n",
       "      nearest_exon_distance   Pool     Hamming_BC  \\\n",
       "0                       NaN  F2-R1  TTATACACAGCTA   \n",
       "1                       NaN  F2-R1  ATGGTGCTCCTGG   \n",
       "2                       NaN  F2-R1  TACGACGAGCGAG   \n",
       "3                       NaN  F2-R1  AGCACGGAAGTGC   \n",
       "4                       NaN  F2-R1  TGGGCCCGCCCAT   \n",
       "...                     ...    ...            ...   \n",
       "32283                   NaN  F2-R6  CCTGTAATCGTTC   \n",
       "32284                   NaN  F2-R6  AGAAACTTACAGC   \n",
       "32285                   NaN  F2-R6  AACAAGGTAACGA   \n",
       "32286                   NaN  F2-R6  TGTGGGATATAAA   \n",
       "32287                   NaN  F2-R6  TTTGACCTAGTTG   \n",
       "\n",
       "                                              gRNA_oligo  polyT RE_sites  \\\n",
       "0      GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCCGCCA...  False        2   \n",
       "1      GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGATGGGG...  False        2   \n",
       "2      GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGGGGCTG...  False        2   \n",
       "3      GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCTGAAG...  False        2   \n",
       "4      GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGAAGGCC...  False        2   \n",
       "...                                                  ...    ...      ...   \n",
       "32283  GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGTTGCGC...  False        2   \n",
       "32284  GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGAAGCGG...  False        2   \n",
       "32285  GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGTGTATG...  False        2   \n",
       "32286  GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGCTACTC...  False        2   \n",
       "32287  GTGTAACCCGTAGGGCACCTAGCGTACACGTCTCACACCGGCACGA...  False        2   \n",
       "\n",
       "      gRNA_id_OLD  \n",
       "0          gRNA_1  \n",
       "1          gRNA_2  \n",
       "2          gRNA_3  \n",
       "3          gRNA_4  \n",
       "4          gRNA_5  \n",
       "...           ...  \n",
       "32283         NaN  \n",
       "32284         NaN  \n",
       "32285         NaN  \n",
       "32286         NaN  \n",
       "32287         NaN  \n",
       "\n",
       "[32288 rows x 27 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_library_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#final sensor check\n",
    "\n",
    "targ1 = combined_library_filtered[combined_library_filtered['classification']!='non-targeting control']\n",
    "\n",
    "for i, val in targ1.iterrows():\n",
    "    proto = val['proto_G+19'][1:]\n",
    "    sensor = str(Bio.Seq.Seq(val['sensor_wt']).reverse_complement())\n",
    "\n",
    "    assert proto in sensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32288\n",
      "31464\n",
      "31630\n",
      "32288\n",
      "32288\n",
      "[250]\n",
      "[2]\n"
     ]
    }
   ],
   "source": [
    "print(len(np.unique(combined_library_filtered['gRNA_oligo'])))\n",
    "print(len(np.unique(combined_library_filtered['proto_G+19'])))\n",
    "print(len(np.unique(combined_library_filtered['sensor_wt'])))\n",
    "print(len(np.unique(combined_library_filtered['gRNA_id'])))\n",
    "print(len(np.unique(combined_library_filtered['Hamming_BC'])))\n",
    "\n",
    "print(np.unique([len(i) for i in combined_library_filtered['gRNA_oligo']]))\n",
    "print(np.unique(combined_library_filtered['RE_sites']))\n",
    "\n",
    "#note that there are duplicate sgRNAs because there are paralogs (most likely)\n",
    "#need to keep this in mind in the future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#label and save\n",
    "combined_library_filtered['species'] = 'Mouse'\n",
    "combined_library_filtered.to_csv('Mouse_tiling_library.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# modfying edit df to exclude filtered guides\n",
    "\n",
    "Haven't done this yet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Men1', 'Kdm6a', 'Kmt2a', 'Kmt2b', 'Kmt2c', 'Kmt2d']"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = list(fgfr2_3b['Gene'])\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "edit_holder2 = []\n",
    "for gene in genes:\n",
    "    targ = combined_library_filtered[(combined_library_filtered['classification']=='targeting') & (combined_library_filtered['Gene']==gene)]\n",
    "    id_dict = dict(zip(targ['gRNA_id_OLD'], targ['gRNA_id']))\n",
    "\n",
    "    e = edit_dict[gene]\n",
    "    e = e[e['gRNA_id'].isin(list(targ['gRNA_id_OLD']))]\n",
    "    e = e.rename(columns = {'gRNA_id':'gRNA_id_OLD'}) \n",
    "\n",
    "    e['gRNA_id'] = [id_dict[i] for i in list(e['gRNA_id_OLD'])]\n",
    "    e['Gene'] = gene\n",
    "\n",
    "    cols = ['gRNA_id', 'edit', 'Editor', 'protospacer', 'exon', 'proto_start',\n",
    "       'proto_end', 'PAM', 'PAM_start', 'PAM_end', 'ideal_start', 'ideal_end',\n",
    "       'ideal_window', 'strand', 'CDS_hit', 'ABE_amenable', 'CBE_amenable',\n",
    "       'HGVSp', 'potential_splice', 'Complex', 'MUT_AA', 'WT_AA', 'Codon',\n",
    "       'sensor_wt', 'sensor_alt', 'Gene','gRNA_id_OLD',]\n",
    "    \n",
    "    e = e[cols]\n",
    "\n",
    "    edit_holder2.append(e)\n",
    "\n",
    "    e.to_csv(f'predicted_edit_tables/mouse_{gene}_predicted_edits.csv', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
